{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Complete Restart with JOGs using Orthofinder results\n",
    "JOGs are Josiah's Orthogroups from Orthofinder run using latest assemblies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HOG file parsing for gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from DNASkittleUtils.Contigs import read_contigs\n",
    "from DNASkittleUtils.DDVUtils import pp\n",
    "from os.path import join, basename, splitext\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def convert_to_gene_id(name):\n",
    "    if name.startswith('FRAX'):  # not FRAEX for excelsior\n",
    "        name = name[name.find('_')+1:]\n",
    "        return name[:name.rfind('_')]  #chop off last _R0 tail\n",
    "    if name.startswith('FRAEX'):\n",
    "        name = name.replace('FRAEX38873_v2_', '')\n",
    "        return name#[:name.rfind('.')]\n",
    "    else:\n",
    "        return name\n",
    "assert convert_to_gene_id(\"FRAX19_000226960.2_R0\") == \"000226960.2\", convert_to_gene_id(\"FRAX19_000226960.2_R0\")\n",
    "assert convert_to_gene_id(\"FRAEX38873_v2_000001220.2\") == \"000001220.2\", convert_to_gene_id(\"FRAEX38873_v2_000001220.2\")\n",
    "tomato = 'Solyc03g095770.2.1 pacid=36135394 transcript=Solyc03g095770.2.1 locus=Solyc03g095770.2 ID=Solyc03g095770.2.1.ITAG2.4 annot-version=ITAG2.4 [Slycopersicum_390_ITAG2]'\n",
    "assert convert_to_gene_id(tomato) == tomato\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def parse_OGs_from_Orthogroups_csv():\n",
    "    Orthogroups_csv = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\Ash_Proteome\\Results_Jun25\\Orthogroups.csv\"\n",
    "    HOGs = {}\n",
    "    with open(Orthogroups_csv, 'r') as csv:\n",
    "        for line in csv:\n",
    "            genes = line.strip().split()\n",
    "            og, genes = genes[0][:-1], genes[1:]  # split off first\n",
    "            HOGs[og] = genes\n",
    "    return HOGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64650"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OGs = parse_OGs_from_Orthogroups_csv()\n",
    "len(OGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['OE6A039394P1']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OGs['OG0047000']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Greedy merge algorithm\n",
    " ( agglomerative clustering? )\n",
    "* Starting clusters can be HOGs\n",
    "* clusters = dict{ gene: pointer to cluster }\n",
    "* Clusters are set(gene)\n",
    "* First gene gets a new cluster\n",
    "* If gene A overlaps with gene B anywhere, then go through each cluster to find gene A and add gene B\n",
    "* If gene B is in any other clusters, find it and merge the entire cluster into gene A cluster\n",
    "* Assert a gene can only ever be in one cluster at a time, except in the atomic operation of merging two\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Starting from HOGs but no side-effects, using networkx to retreive clusters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[Python Network Graphs](https://www.python.org/doc/essays/graphs/)\n",
    "\n",
    "[Creating a networkx graph](https://networkx.github.io/documentation/networkx-1.10/tutorial/tutorial.html)  \n",
    "[Graph Connected Components](https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.algorithms.components.connected.connected_components.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'gene3': {'HOG2'}, 'gene1': {'HOG1'}, 'gene2': {'HOG2', 'HOG1'}} \n",
      " {'gene3': {'HOG2'}, 'gene1': {'HOG1'}, 'gene2': {'HOG2', 'HOG1'}}\n"
     ]
    }
   ],
   "source": [
    "def map_genes_to_HOGs(HOGs):\n",
    "    genes_to_HOGs = {}\n",
    "    for hog, genes in HOGs.items():\n",
    "        for gene in genes:\n",
    "            if gene not in genes_to_HOGs:\n",
    "                genes_to_HOGs[gene] = set()\n",
    "            genes_to_HOGs[gene].add(hog)\n",
    "    return genes_to_HOGs\n",
    "\n",
    "def test_map_genes_to_HOGs():\n",
    "    genes_to_HOGs_answer = {'gene1': {'HOG1'},\n",
    "                     'gene2': {'HOG1', 'HOG2'},\n",
    "                     'gene3': {'HOG2'}}\n",
    "    HOGs = {'HOG1': set(('gene1', 'gene2')),\n",
    "            'HOG2': set(('gene2', 'gene3'))}  # all HOGs start out parsed and non-exlusive\n",
    "    genes_to_HOGs = map_genes_to_HOGs(HOGs)\n",
    "    print(genes_to_HOGs, '\\n',genes_to_HOGs_answer)\n",
    "    assert genes_to_HOGs == genes_to_HOGs_answer\n",
    "test_map_genes_to_HOGs()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from itertools import combinations\n",
    "import networkx\n",
    "\n",
    "def create_cluster_network(genes_to_HOGs, HOGs, blacklist_genes=set(), blacklist_HOGs=set()):\n",
    "    g = networkx.Graph()\n",
    "    g.add_nodes_from(HOGs.keys())\n",
    "    for gene, hogs in genes_to_HOGs.items():\n",
    "        if len(hogs) > 1:\n",
    "            if gene not in blacklist_genes:\n",
    "                target_hogs = hogs.difference(blacklist_HOGs)\n",
    "                # two HOGs have overlap and need to be merged\n",
    "                g.add_edges_from(combinations(target_hogs, 2))  # add edge between all hogs\n",
    "            \n",
    "    return g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'HOG2-HOG1': {'gene3', 'gene1', 'gene2'}}\n"
     ]
    }
   ],
   "source": [
    "def make_super_HOGs(HOGs, networked_hogs):\n",
    "    \"\"\"collect subnetworks together into larger clusters\"\"\"\n",
    "    super_HOGs = {}\n",
    "    genes_seen = set()\n",
    "    clusters = [c for c in sorted(networkx.connected_components(networked_hogs), key=len, reverse=True)]\n",
    "    for cluster in clusters:\n",
    "        name = '-'.join(cluster)\n",
    "        super_HOGs[name] = set().union(gene for hog in cluster for gene in HOGs[hog])\n",
    "        assert super_HOGs[name] not in genes_seen, \"You missed a clustering connection.  Genes should only occur once\"\n",
    "        genes_seen.update(super_HOGs[name])\n",
    "            \n",
    "    return super_HOGs\n",
    "\n",
    "def test_make_super_HOGs():\n",
    "    super_HOGs_answer = {'HOG2-HOG1': {'gene1', 'gene2', 'gene3'}}\n",
    "    HOGs = {'HOG1': set(('gene1', 'gene2')),\n",
    "            'HOG2': set(('gene2', 'gene3'))}  # all HOGs start out parsed and non-exlusive\n",
    "    genes_to_HOGs = map_genes_to_HOGs(HOGs)\n",
    "    network = create_cluster_network(genes_to_HOGs, HOGs)\n",
    "    assert str(network.edges()) == \"[('HOG2', 'HOG1')]\", network.edges()\n",
    "    super_HOGs = make_super_HOGs(HOGs, network)\n",
    "    print(super_HOGs)\n",
    "    assert super_HOGs == super_HOGs_answer, super_HOGs\n",
    "test_make_super_HOGs()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### New way of clustering and tabulating using only orthogroups.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ogs_to_specific_genes = parse_OGs_from_Orthogroups_csv()\n",
    "ogs_to_generic_genes = {og: set(convert_to_gene_id(g) for g in genes) for og, genes in ogs_to_specific_genes.items()}\n",
    "generic_genes_to_ogs = map_genes_to_HOGs(ogs_to_generic_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promiscuous_genes = [gene for gene, hogs in generic_genes_to_ogs.items() if len(hogs) > 20]\n",
    "print(len(promiscuous_genes))\n",
    "promiscuous_OGs = [hog for hog, genes in ogs_to_specific_genes.items() if len(genes) > 350]\n",
    "len(promiscuous_OGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['OG0000052',\n",
       " 'OG0000058',\n",
       " 'OG0000071',\n",
       " 'OG0000013',\n",
       " 'OG0000042',\n",
       " 'OG0000066',\n",
       " 'OG0000063',\n",
       " 'OG0000007',\n",
       " 'OG0000047',\n",
       " 'OG0000070',\n",
       " 'OG0000036',\n",
       " 'OG0000065',\n",
       " 'OG0000005',\n",
       " 'OG0000009',\n",
       " 'OG0000075',\n",
       " 'OG0000014',\n",
       " 'OG0000022',\n",
       " 'OG0000051',\n",
       " 'OG0000027',\n",
       " 'OG0000015',\n",
       " 'OG0000043',\n",
       " 'OG0000060',\n",
       " 'OG0000006',\n",
       " 'OG0000050',\n",
       " 'OG0000021',\n",
       " 'OG0000038',\n",
       " 'OG0000037',\n",
       " 'OG0000002',\n",
       " 'OG0000016',\n",
       " 'OG0000069',\n",
       " 'OG0000053',\n",
       " 'OG0000010',\n",
       " 'OG0000011',\n",
       " 'OG0000018',\n",
       " 'OG0000001',\n",
       " 'OG0000057',\n",
       " 'OG0000049',\n",
       " 'OG0000028',\n",
       " 'OG0000045',\n",
       " 'OG0000029',\n",
       " 'OG0000059',\n",
       " 'OG0000041',\n",
       " 'OG0000056',\n",
       " 'OG0000003',\n",
       " 'OG0000068',\n",
       " 'OG0000019',\n",
       " 'OG0000012',\n",
       " 'OG0000023',\n",
       " 'OG0000030',\n",
       " 'OG0000033',\n",
       " 'OG0000064',\n",
       " 'OG0000055',\n",
       " 'OG0000020',\n",
       " 'OG0000039',\n",
       " 'OG0000048',\n",
       " 'OG0000017',\n",
       " 'OG0000073',\n",
       " 'OG0000040',\n",
       " 'OG0000008',\n",
       " 'OG0000054',\n",
       " 'OG0000074',\n",
       " 'OG0000032',\n",
       " 'OG0000044',\n",
       " 'OG0000035',\n",
       " 'OG0000061',\n",
       " 'OG0000024',\n",
       " 'OG0000067',\n",
       " 'OG0000004',\n",
       " 'OG0000000',\n",
       " 'OG0000034',\n",
       " 'OG0000031',\n",
       " 'OG0000072',\n",
       " 'OG0000062',\n",
       " 'OG0000026',\n",
       " 'OG0000025',\n",
       " 'OG0000046']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "promiscuous_OGs\n",
    "\n",
    "ogs_to_specific_genes['OG0000052']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "network = create_cluster_network(generic_genes_to_ogs, ogs_to_generic_genes, promiscuous_genes, promiscuous_OGs)\n",
    "JSG_to_specific_genes = make_super_HOGs(ogs_to_specific_genes, network)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'44,018'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pp(len(JSG_to_specific_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "all_species = 'FRAX00\tFRAX01\tFRAX02\tFRAX03\tFRAX04\tFRAX05\tFRAX06\tFRAX07\tFRAX08\tFRAX09\tFRAX11\tFRAX12\tFRAX13\tFRAX14\tFRAX15\tFRAX16\tFRAX19\tFRAX20\tFRAX21\tFRAX23\tFRAX25\tFRAX26\tFRAX27\tFRAX28\tFRAX29\tFRAX30\tFRAX31\tFRAX32\tFRAX33\tFRAX34\tOeuropea\tMguttatus\tSlycopersicum'.split('\\t')\n",
    "species_aliases = {'OE6': 'Oeuropea', 'Sol': 'Slycopersicum', 'Mig': 'Mguttatus'}\n",
    "def extract_species(gene_name):\n",
    "    if gene_name.startswith('FRAEX'):\n",
    "        return 'FRAX00'\n",
    "    elif gene_name.startswith('FRAX'):\n",
    "        return gene_name.split('_')[0]\n",
    "    else:\n",
    "        three = gene_name[:3]\n",
    "        return species_aliases[three]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44018\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Desc</th>\n",
       "      <th>FRAX00</th>\n",
       "      <th>FRAX01</th>\n",
       "      <th>FRAX02</th>\n",
       "      <th>FRAX03</th>\n",
       "      <th>FRAX04</th>\n",
       "      <th>FRAX05</th>\n",
       "      <th>FRAX06</th>\n",
       "      <th>FRAX07</th>\n",
       "      <th>FRAX08</th>\n",
       "      <th>...</th>\n",
       "      <th>FRAX28</th>\n",
       "      <th>FRAX29</th>\n",
       "      <th>FRAX30</th>\n",
       "      <th>FRAX31</th>\n",
       "      <th>FRAX32</th>\n",
       "      <th>FRAX33</th>\n",
       "      <th>FRAX34</th>\n",
       "      <th>Oeuropea</th>\n",
       "      <th>Mguttatus</th>\n",
       "      <th>Slycopersicum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>JSG00000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JSG00001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JSG00002</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JSG00003</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JSG00004</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Desc  FRAX00  FRAX01  FRAX02  FRAX03  FRAX04  FRAX05  FRAX06  FRAX07  \\\n",
       "0  JSG00000       0       0       0       0       0       0       0       0   \n",
       "1  JSG00001       0       0       0       0       0       0       0       0   \n",
       "2  JSG00002       2       1       6       3       2       3       2       2   \n",
       "3  JSG00003       1       3       2       1       1       0       2       1   \n",
       "4  JSG00004       0       0       0       0       0       0       0       0   \n",
       "\n",
       "   FRAX08      ...        FRAX28  FRAX29  FRAX30  FRAX31  FRAX32  FRAX33  \\\n",
       "0       0      ...             0       0       0       0       0       0   \n",
       "1       0      ...             0       0       0       0       0       0   \n",
       "2       4      ...             3       4       3       2       4       3   \n",
       "3       1      ...             1       1       1       1       1       1   \n",
       "4       0      ...             0       0       0       0       0       0   \n",
       "\n",
       "   FRAX34  Oeuropea  Mguttatus  Slycopersicum  \n",
       "0       0         0          0              0  \n",
       "1       0         0          1              0  \n",
       "2       3         3          1              2  \n",
       "3       0         2          1              1  \n",
       "4       0         0          1              0  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pandas import DataFrame    \n",
    "def create_count_table_from_JSGs(JSG_to_specific_genes):\n",
    "    \"\"\"Map generic names back to their specifics and perform species count table based on those names\"\"\"\n",
    "    table = []\n",
    "    for jsg, specific_genes in JSG_to_specific_genes.items():\n",
    "        species_row = {species: 0 for species in all_species}\n",
    "        species_row['Desc'] = 'JSG' + str(len(table)).zfill(5)\n",
    "        for gene in specific_genes:\n",
    "            species_row[ extract_species(gene) ] += 1\n",
    "        table.append(species_row)\n",
    "        \n",
    "    df = DataFrame(table, columns=['Desc'] + all_species) #table.loc[species_row['Family ID']] = species_row\n",
    "    return df\n",
    "\n",
    "table = create_count_table_from_JSGs(JSG_to_specific_genes)\n",
    "print(len(table))\n",
    "table.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Output new CSV counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "with open('Ash_Proteome/Results_Jun25/JSG_to_OG_mappings.csv', 'w') as out:\n",
    "    line_count = 0\n",
    "    for jsg, specific_genes in JSG_to_specific_genes.items(): \n",
    "        name = 'JSG' + str(line_count).zfill(5)\n",
    "        line_count += 1\n",
    "        out.write(name + \"\\t\" + ','.join(jsg.split('-')) + '\\t' + ','.join(specific_genes) + '\\n' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Desc</th>\n",
       "      <th>sum</th>\n",
       "      <th>FRAX00</th>\n",
       "      <th>FRAX01</th>\n",
       "      <th>FRAX02</th>\n",
       "      <th>FRAX03</th>\n",
       "      <th>FRAX04</th>\n",
       "      <th>FRAX05</th>\n",
       "      <th>FRAX06</th>\n",
       "      <th>FRAX07</th>\n",
       "      <th>...</th>\n",
       "      <th>FRAX28</th>\n",
       "      <th>FRAX29</th>\n",
       "      <th>FRAX30</th>\n",
       "      <th>FRAX31</th>\n",
       "      <th>FRAX32</th>\n",
       "      <th>FRAX33</th>\n",
       "      <th>FRAX34</th>\n",
       "      <th>Oeuropea</th>\n",
       "      <th>Mguttatus</th>\n",
       "      <th>Slycopersicum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Family ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14444</th>\n",
       "      <td>JSG14444</td>\n",
       "      <td>3349</td>\n",
       "      <td>110</td>\n",
       "      <td>108</td>\n",
       "      <td>109</td>\n",
       "      <td>105</td>\n",
       "      <td>90</td>\n",
       "      <td>105</td>\n",
       "      <td>95</td>\n",
       "      <td>108</td>\n",
       "      <td>...</td>\n",
       "      <td>110</td>\n",
       "      <td>109</td>\n",
       "      <td>105</td>\n",
       "      <td>69</td>\n",
       "      <td>110</td>\n",
       "      <td>110</td>\n",
       "      <td>84</td>\n",
       "      <td>209</td>\n",
       "      <td>74</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28342</th>\n",
       "      <td>JSG28342</td>\n",
       "      <td>2797</td>\n",
       "      <td>22</td>\n",
       "      <td>106</td>\n",
       "      <td>84</td>\n",
       "      <td>93</td>\n",
       "      <td>79</td>\n",
       "      <td>114</td>\n",
       "      <td>100</td>\n",
       "      <td>102</td>\n",
       "      <td>...</td>\n",
       "      <td>93</td>\n",
       "      <td>107</td>\n",
       "      <td>94</td>\n",
       "      <td>84</td>\n",
       "      <td>98</td>\n",
       "      <td>105</td>\n",
       "      <td>85</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40369</th>\n",
       "      <td>JSG40369</td>\n",
       "      <td>2551</td>\n",
       "      <td>79</td>\n",
       "      <td>78</td>\n",
       "      <td>89</td>\n",
       "      <td>77</td>\n",
       "      <td>74</td>\n",
       "      <td>82</td>\n",
       "      <td>75</td>\n",
       "      <td>82</td>\n",
       "      <td>...</td>\n",
       "      <td>79</td>\n",
       "      <td>85</td>\n",
       "      <td>75</td>\n",
       "      <td>78</td>\n",
       "      <td>81</td>\n",
       "      <td>90</td>\n",
       "      <td>67</td>\n",
       "      <td>102</td>\n",
       "      <td>54</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32193</th>\n",
       "      <td>JSG32193</td>\n",
       "      <td>2294</td>\n",
       "      <td>76</td>\n",
       "      <td>67</td>\n",
       "      <td>69</td>\n",
       "      <td>67</td>\n",
       "      <td>66</td>\n",
       "      <td>68</td>\n",
       "      <td>72</td>\n",
       "      <td>71</td>\n",
       "      <td>...</td>\n",
       "      <td>68</td>\n",
       "      <td>74</td>\n",
       "      <td>66</td>\n",
       "      <td>59</td>\n",
       "      <td>66</td>\n",
       "      <td>69</td>\n",
       "      <td>61</td>\n",
       "      <td>109</td>\n",
       "      <td>89</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2784</th>\n",
       "      <td>JSG02784</td>\n",
       "      <td>2156</td>\n",
       "      <td>49</td>\n",
       "      <td>64</td>\n",
       "      <td>65</td>\n",
       "      <td>59</td>\n",
       "      <td>54</td>\n",
       "      <td>73</td>\n",
       "      <td>72</td>\n",
       "      <td>79</td>\n",
       "      <td>...</td>\n",
       "      <td>75</td>\n",
       "      <td>71</td>\n",
       "      <td>76</td>\n",
       "      <td>54</td>\n",
       "      <td>66</td>\n",
       "      <td>65</td>\n",
       "      <td>52</td>\n",
       "      <td>200</td>\n",
       "      <td>43</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Desc   sum  FRAX00  FRAX01  FRAX02  FRAX03  FRAX04  FRAX05  \\\n",
       "Family ID                                                                   \n",
       "14444      JSG14444  3349     110     108     109     105      90     105   \n",
       "28342      JSG28342  2797      22     106      84      93      79     114   \n",
       "40369      JSG40369  2551      79      78      89      77      74      82   \n",
       "32193      JSG32193  2294      76      67      69      67      66      68   \n",
       "2784       JSG02784  2156      49      64      65      59      54      73   \n",
       "\n",
       "           FRAX06  FRAX07      ...        FRAX28  FRAX29  FRAX30  FRAX31  \\\n",
       "Family ID                      ...                                         \n",
       "14444          95     108      ...           110     109     105      69   \n",
       "28342         100     102      ...            93     107      94      84   \n",
       "40369          75      82      ...            79      85      75      78   \n",
       "32193          72      71      ...            68      74      66      59   \n",
       "2784           72      79      ...            75      71      76      54   \n",
       "\n",
       "           FRAX32  FRAX33  FRAX34  Oeuropea  Mguttatus  Slycopersicum  \n",
       "Family ID                                                              \n",
       "14444         110     110      84       209         74             75  \n",
       "28342          98     105      85        23          0              0  \n",
       "40369          81      90      67       102         54             48  \n",
       "32193          66      69      61       109         89             75  \n",
       "2784           66      65      52       200         43             15  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sums = table.sum(numeric_only=True, axis=1)\n",
    "df = table.assign(sum= sums)\n",
    "columns=['Desc', 'sum'] + all_species\n",
    "df.index.name = 'Family ID'\n",
    "df = df[columns]\n",
    "df.sort_values('sum', ascending=False, inplace=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df.to_csv('Ash_Proteome/Results_Jun25/JSG_GeneCounts.csv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Desc</th>\n",
       "      <th>sum</th>\n",
       "      <th>FRAX00</th>\n",
       "      <th>FRAX01</th>\n",
       "      <th>FRAX02</th>\n",
       "      <th>FRAX03</th>\n",
       "      <th>FRAX04</th>\n",
       "      <th>FRAX05</th>\n",
       "      <th>FRAX06</th>\n",
       "      <th>FRAX07</th>\n",
       "      <th>...</th>\n",
       "      <th>FRAX28</th>\n",
       "      <th>FRAX29</th>\n",
       "      <th>FRAX30</th>\n",
       "      <th>FRAX31</th>\n",
       "      <th>FRAX32</th>\n",
       "      <th>FRAX33</th>\n",
       "      <th>FRAX34</th>\n",
       "      <th>Oeuropea</th>\n",
       "      <th>Mguttatus</th>\n",
       "      <th>Slycopersicum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Family ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22517</th>\n",
       "      <td>JSG22517</td>\n",
       "      <td>26</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Desc  sum  FRAX00  FRAX01  FRAX02  FRAX03  FRAX04  FRAX05  \\\n",
       "Family ID                                                                  \n",
       "22517      JSG22517   26       1       1       1       0       1       1   \n",
       "\n",
       "           FRAX06  FRAX07      ...        FRAX28  FRAX29  FRAX30  FRAX31  \\\n",
       "Family ID                      ...                                         \n",
       "22517           1       1      ...             1       1       1       1   \n",
       "\n",
       "           FRAX32  FRAX33  FRAX34  Oeuropea  Mguttatus  Slycopersicum  \n",
       "Family ID                                                              \n",
       "22517           0       0       2         1          0              1  \n",
       "\n",
       "[1 rows x 35 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['Desc'] == 'JSG22517']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'FRAX00,FRAX01,FRAX02,FRAX03,FRAX04,FRAX05,FRAX06,FRAX07,FRAX08,FRAX09,FRAX11,FRAX12,FRAX13,FRAX14,FRAX15,FRAX16,FRAX19,FRAX20,FRAX21,FRAX23,FRAX25,FRAX26,FRAX27,FRAX28,FRAX29,FRAX30,FRAX31,FRAX32,FRAX33,FRAX34,Oeuropea,Mguttatus,Slycopersicum'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "','.join(all_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fraxinus_columns = df.loc[: ,'FRAX00':'FRAX34']\n",
    "df['Fraxinus Std Dev'] = fraxinus_columns.std(axis=1)\n",
    "df['Fraxinus Avg'] = fraxinus_columns.mean(axis=1)\n",
    "df['Fraxinus Max'] = fraxinus_columns.max(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df['Fraxinus Min'] = fraxinus_columns.min(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44018"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14696"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "criteria = (df['Fraxinus Avg'] > 1/15) & (df['Fraxinus Std Dev'] > 0)\n",
    "criteria.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6653"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "criteria = (df['Fraxinus Avg'] > 1/15) & (df['Fraxinus Std Dev'] > 0) & (df['Fraxinus Max'] <= 2 )\n",
    "criteria.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1319"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "criteria = (df['Fraxinus Avg'] > 1/15) & (df['Fraxinus Std Dev'] > 0) & (df['Fraxinus Max'] <= 2 ) & (df['Fraxinus Min'] == 1)\n",
    "criteria.sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of 14,696 we lose about half 6,653 to large families.  That's a usable amount.  Eliminating 0's drops it to 1319, which probably isn't worth it.  Trinary states isn't so bad."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for jsg in df[criteria]['Desc']:\n",
    "    # we immediately run into an inability to map back to OGs\n",
    "    # I need to determine JSG names and save the mapping explicitly in make_super_HOGs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Everything that follows is Old method using GeneCount.csv file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_________"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Actual Clusters with Real HOGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'13,306  Reduced count'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "actual_genes_to_OGs = map_genes_to_HOGs(OGs)\n",
    "actual_network = create_cluster_network(actual_genes_to_OGs, OGs)\n",
    "super_OGs = make_super_HOGs(OGs, actual_network)\n",
    "\n",
    "\"{:,}\".format(len(OGs) - len(super_OGs)) + \"  Reduced count\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compose new alignment with Super HOGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'5,347'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from DNASkittleUtils.DDVUtils import pp\n",
    "modified_HOGs_only = [c for c in sorted(networkx.connected_components(actual_network), key=len, reverse=True) if len(c) > 1]\n",
    "pp(len(modified_HOGs_only))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "assert all(['-'.join(h) in super_OGs for h in modified_HOGs_only])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Blacklist promiscuous genes and HOGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "actual_genes_to_OGs = map_genes_to_HOGs(OGs)\n",
    "len(actual_genes_to_OGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "promiscuous_genes = [gene for gene, hogs in actual_genes_to_OGs.items() if len(hogs) > 5]\n",
    "len(promiscuous_genes)\n",
    "\n",
    "promiscuous_OGs = [hog for hog, genes in OGs.items() if len(genes) > 150]\n",
    "len(promiscuous_OGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11,713  Reduced count\n"
     ]
    }
   ],
   "source": [
    "actual_network = create_cluster_network(actual_genes_to_OGs, OGs, promiscuous_genes, promiscuous_OGs)\n",
    "super_HOGs = make_super_HOGs(OGs, actual_network)\n",
    "\n",
    "print(\"{:,}\".format(len(OGs) - len(super_HOGs)) + \"  Reduced count\")\n",
    "smart_modified_HOGs = [c for c in sorted(networkx.connected_components(actual_network), key=len, reverse=True) if len(c) > 1]\n",
    "pp(len(smart_modified_HOGs)), \"HOGs\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "52937"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(super_HOGs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cluster rows by new super groups and output GeneCount.csv for new CAFE run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'OG0000848',\n",
       "  'OG0001316',\n",
       "  'OG0001331',\n",
       "  'OG0001620',\n",
       "  'OG0001679',\n",
       "  'OG0002719',\n",
       "  'OG0003830',\n",
       "  'OG0005353',\n",
       "  'OG0007285',\n",
       "  'OG0007534',\n",
       "  'OG0008375',\n",
       "  'OG0009430',\n",
       "  'OG0010137',\n",
       "  'OG0010197',\n",
       "  'OG0010551',\n",
       "  'OG0011854',\n",
       "  'OG0012187',\n",
       "  'OG0012608',\n",
       "  'OG0014503',\n",
       "  'OG0018028',\n",
       "  'OG0019062',\n",
       "  'OG0019438',\n",
       "  'OG0020872',\n",
       "  'OG0023528',\n",
       "  'OG0023682',\n",
       "  'OG0025173',\n",
       "  'OG0031433',\n",
       "  'OG0031546',\n",
       "  'OG0036452'},\n",
       " {'OG0000845',\n",
       "  'OG0002526',\n",
       "  'OG0002707',\n",
       "  'OG0002935',\n",
       "  'OG0006075',\n",
       "  'OG0008106',\n",
       "  'OG0008524',\n",
       "  'OG0009022',\n",
       "  'OG0016155',\n",
       "  'OG0019891',\n",
       "  'OG0020765',\n",
       "  'OG0022144',\n",
       "  'OG0024521',\n",
       "  'OG0025822',\n",
       "  'OG0026320',\n",
       "  'OG0027131',\n",
       "  'OG0027770',\n",
       "  'OG0030545',\n",
       "  'OG0030884',\n",
       "  'OG0031134',\n",
       "  'OG0032084',\n",
       "  'OG0032507',\n",
       "  'OG0032861',\n",
       "  'OG0032996',\n",
       "  'OG0033679',\n",
       "  'OG0035244',\n",
       "  'OG0035659',\n",
       "  'OG0037245',\n",
       "  'OG0038070'},\n",
       " {'OG0017549',\n",
       "  'OG0019437',\n",
       "  'OG0019487',\n",
       "  'OG0019883',\n",
       "  'OG0020636',\n",
       "  'OG0020785',\n",
       "  'OG0024164',\n",
       "  'OG0030401',\n",
       "  'OG0030670',\n",
       "  'OG0031104',\n",
       "  'OG0031668',\n",
       "  'OG0032242',\n",
       "  'OG0032459',\n",
       "  'OG0033742',\n",
       "  'OG0035597',\n",
       "  'OG0036261',\n",
       "  'OG0036639',\n",
       "  'OG0037433',\n",
       "  'OG0037991',\n",
       "  'OG0038268',\n",
       "  'OG0038313',\n",
       "  'OG0038342',\n",
       "  'OG0038353'},\n",
       " {'OG0000530',\n",
       "  'OG0000638',\n",
       "  'OG0002119',\n",
       "  'OG0002383',\n",
       "  'OG0005498',\n",
       "  'OG0007610',\n",
       "  'OG0007752',\n",
       "  'OG0007928',\n",
       "  'OG0008499',\n",
       "  'OG0009098',\n",
       "  'OG0012063',\n",
       "  'OG0014564',\n",
       "  'OG0017804',\n",
       "  'OG0018060',\n",
       "  'OG0019108',\n",
       "  'OG0019469',\n",
       "  'OG0022648',\n",
       "  'OG0030512',\n",
       "  'OG0033296',\n",
       "  'OG0035940',\n",
       "  'OG0036324',\n",
       "  'OG0036696'},\n",
       " {'OG0000870',\n",
       "  'OG0009632',\n",
       "  'OG0017265',\n",
       "  'OG0020544',\n",
       "  'OG0022579',\n",
       "  'OG0024491',\n",
       "  'OG0024696',\n",
       "  'OG0026225',\n",
       "  'OG0027389',\n",
       "  'OG0030388',\n",
       "  'OG0030515',\n",
       "  'OG0031500',\n",
       "  'OG0031596',\n",
       "  'OG0031931',\n",
       "  'OG0034689',\n",
       "  'OG0034779',\n",
       "  'OG0034990',\n",
       "  'OG0035796',\n",
       "  'OG0037917',\n",
       "  'OG0038486',\n",
       "  'OG0038513'},\n",
       " {'OG0000906',\n",
       "  'OG0001413',\n",
       "  'OG0001854',\n",
       "  'OG0002521',\n",
       "  'OG0007836',\n",
       "  'OG0008682',\n",
       "  'OG0010326',\n",
       "  'OG0013655',\n",
       "  'OG0013793',\n",
       "  'OG0015422',\n",
       "  'OG0016510',\n",
       "  'OG0016571',\n",
       "  'OG0019076',\n",
       "  'OG0019473',\n",
       "  'OG0020847',\n",
       "  'OG0023683',\n",
       "  'OG0024774',\n",
       "  'OG0025619',\n",
       "  'OG0025627',\n",
       "  'OG0033307',\n",
       "  'OG0036725'},\n",
       " {'OG0000430',\n",
       "  'OG0018510',\n",
       "  'OG0020598',\n",
       "  'OG0021473',\n",
       "  'OG0023218',\n",
       "  'OG0023613',\n",
       "  'OG0024915',\n",
       "  'OG0025302',\n",
       "  'OG0026863',\n",
       "  'OG0027449',\n",
       "  'OG0027674',\n",
       "  'OG0027828',\n",
       "  'OG0031922',\n",
       "  'OG0033441',\n",
       "  'OG0033953',\n",
       "  'OG0034071',\n",
       "  'OG0036293',\n",
       "  'OG0037757',\n",
       "  'OG0037906',\n",
       "  'OG0037974'},\n",
       " {'OG0011930',\n",
       "  'OG0016986',\n",
       "  'OG0017575',\n",
       "  'OG0017934',\n",
       "  'OG0019996',\n",
       "  'OG0021109',\n",
       "  'OG0021260',\n",
       "  'OG0024334',\n",
       "  'OG0027095',\n",
       "  'OG0031016',\n",
       "  'OG0031116',\n",
       "  'OG0031352',\n",
       "  'OG0032275',\n",
       "  'OG0032501',\n",
       "  'OG0033324',\n",
       "  'OG0033696',\n",
       "  'OG0035446',\n",
       "  'OG0038311',\n",
       "  'OG0038354'},\n",
       " {'OG0000588',\n",
       "  'OG0000917',\n",
       "  'OG0001434',\n",
       "  'OG0001638',\n",
       "  'OG0002567',\n",
       "  'OG0002762',\n",
       "  'OG0003036',\n",
       "  'OG0005173',\n",
       "  'OG0005591',\n",
       "  'OG0010459',\n",
       "  'OG0010973',\n",
       "  'OG0012795',\n",
       "  'OG0013752',\n",
       "  'OG0016721',\n",
       "  'OG0017632',\n",
       "  'OG0025912',\n",
       "  'OG0032296',\n",
       "  'OG0034242',\n",
       "  'OG0037434'},\n",
       " {'OG0000817',\n",
       "  'OG0000877',\n",
       "  'OG0002397',\n",
       "  'OG0003603',\n",
       "  'OG0007668',\n",
       "  'OG0019640',\n",
       "  'OG0026371',\n",
       "  'OG0026971',\n",
       "  'OG0027786',\n",
       "  'OG0029565',\n",
       "  'OG0032718',\n",
       "  'OG0032798',\n",
       "  'OG0033337',\n",
       "  'OG0033639',\n",
       "  'OG0035852',\n",
       "  'OG0038083',\n",
       "  'OG0038328'}]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "smart_modified_HOGs[:10]  # going to be used for grouping rows together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Desc</th>\n",
       "      <th>Family ID</th>\n",
       "      <th>FRAX00</th>\n",
       "      <th>FRAX01</th>\n",
       "      <th>FRAX02</th>\n",
       "      <th>FRAX03</th>\n",
       "      <th>FRAX04</th>\n",
       "      <th>FRAX05</th>\n",
       "      <th>FRAX06</th>\n",
       "      <th>FRAX07</th>\n",
       "      <th>...</th>\n",
       "      <th>FRAX28</th>\n",
       "      <th>FRAX29</th>\n",
       "      <th>FRAX30</th>\n",
       "      <th>FRAX31</th>\n",
       "      <th>FRAX32</th>\n",
       "      <th>FRAX33</th>\n",
       "      <th>FRAX34</th>\n",
       "      <th>Mguttatus</th>\n",
       "      <th>Oeuropea</th>\n",
       "      <th>Slycopersicum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>22</td>\n",
       "      <td>29</td>\n",
       "      <td>17</td>\n",
       "      <td>35</td>\n",
       "      <td>23</td>\n",
       "      <td>28</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>22</td>\n",
       "      <td>28</td>\n",
       "      <td>24</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>34</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Desc  Family ID  FRAX00  FRAX01  FRAX02  FRAX03  FRAX04  FRAX05  \\\n",
       "0  OG0000000          0      11      22      29      17      35      23   \n",
       "1  OG0000001          1      22      20      15      19      20      16   \n",
       "2  OG0000002          2      19      15      12      15      16      17   \n",
       "3  OG0000003          3      24      10       6      12      15      13   \n",
       "4  OG0000004          4      20      16      11      10      13      15   \n",
       "\n",
       "   FRAX06  FRAX07      ...        FRAX28  FRAX29  FRAX30  FRAX31  FRAX32  \\\n",
       "0      28      23      ...            27      27      22      28      24   \n",
       "1      20      17      ...            16      16      20      20      18   \n",
       "2      18      17      ...            17      15      17      15      13   \n",
       "3      15      17      ...            13      19      12      15      14   \n",
       "4      16      15      ...            13      14      15      15      14   \n",
       "\n",
       "   FRAX33  FRAX34  Mguttatus  Oeuropea  Slycopersicum  \n",
       "0      23      20          1        19              5  \n",
       "1      20      20          9        20             10  \n",
       "2      18      18          9        22             13  \n",
       "3      13      12         21        34             24  \n",
       "4      15      15          8        13             12  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pandas import read_csv\n",
    "counts = read_csv(r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\Ash_Proteome\\Results_Jun25\\Cafe_orthofinder_Orthogroups.GeneCount.csv\", sep='\\t')\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33693"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each group of OGs:\n",
    "- sum their rows\n",
    "- add a new row called JSG0028362+\n",
    "- delete previously used rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "new_table = counts.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29 26; 29 17; 23 7; 22 17; 21 9; 21 19; 20 12; 19 9; 19 16; 17 9; 17 10; 17 9; 17 4; 17 13; 17 10; 17 5; 17 9; 16 5; 16 9; 16 11; 16 12; 15 13; 15 13; 15 9; 15 5; 15 6; 15 6; 14 10; 14 7; 14 10; 14 7; 14 12; 14 5; 14 11; 14 10; 14 11; 14 13; 13 8; 13 5; 13 12; 13 7; 13 11; 13 9; 13 8; 13 3; 13 3; 13 9; 13 8; 13 7; 13 4; 13 9; 13 6; 13 11; 13 7; 13 3; 13 9; 13 6; 12 10; 12 6; 12 9; 12 6; 12 8; 12 9; 12 1; 12 7; 12 7; 12 11; 12 11; 12 8; 12 10; 12 8; 12 10; 12 7; 12 7; 11 9; 11 6; 11 7; 11 5; 11 5; 11 7; 11 7; 11 8; 11 7; 11 10; 11 10; 11 9; 11 3; 11 2; 11 6; 11 5; 11 9; 11 10; 11 1; 10 4; 10 9; 10 2; 10 9; 10 4; 10 6; 10 9; 10 3; 10 5; 10 6; 10 3; 10 7; 10 9; 10 6; 10 9; 10 2; 10 5; 10 6; 10 6; 10 6; 10 5; 10 6; 10 3; 10 2; 10 5; 10 9; 10 6; 10 9; 9 8; 9 7; 9 7; 9 4; 9 4; 9 8; 9 3; 9 4; 9 8; 9 5; 9 8; 9 6; 9 7; 9 8; 9 6; 9 6; 9 5; 9 4; 9 8; 9 6; 9 6; 9 7; 9 5; 9 3; 9 7; 9 7; 9 6; 9 7; 9 6; 9 5; 9 6; 9 2; 9 7; 9 5; 9 4; 9 3; 9 4; 9 6; 9 5; 9 5; 9 4; 9 3; 9 4; 9 5; 9 4; 9 3; 9 3; 9 7; 9 8; 9 1; 9 6; 8 6; 8 5; 8 4; 8 5; 8 2; 8 7; 8 7; 8 5; 8 1; 8 5; 8 5; 8 5; 8 5; 8 5; 8 6; 8 5; 8 3; 8 7; 8 5; 8 5; 8 2; 8 7; 8 6; 8 5; 8 7; 8 6; 8 4; 8 7; 8 6; 8 7; 8 5; 8 6; 8 3; 8 7; 8 5; 8 7; 8 5; 8 3; 8 6; 8 6; 8 6; 8 2; 8 5; 8 3; 8 6; 8 7; 8 7; 8 5; 8 7; 8 6; 8 7; 8 2; 8 4; 8 7; 8 7; 8 7; 8 6; 8 6; 8 5; 8 1; 8 5; 8 3; 8 6; 8 7; 8 3; 8 4; 8 5; 8 5; 7 6; 7 4; 7 4; 7 5; 7 3; 7 6; 7 3; 7 6; 7 6; 7 6; 7 1; 7 4; 7 3; 7 4; 7 6; 7 5; 7 5; 7 6; 7 5; 7 2; 7 4; 7 4; 7 4; 7 3; 7 6; 7 4; 7 3; 7 5; 7 6; 7 6; 7 4; 7 3; 7 3; 7 5; 7 5; 7 5; 7 5; 7 5; 7 3; 7 4; 7 5; 7 5; 7 2; 7 5; 7 6; 7 6; 7 4; 7 3; 7 3; 7 5; 7 6; 7 3; 7 5; 7 3; 7 6; 7 5; 7 6; 7 2; 7 4; 7 6; 7 4; 7 5; 7 5; 7 3; 7 6; 7 6; 7 4; 7 4; 7 4; 7 5; 7 5; 7 4; 7 2; 7 4; 7 3; 7 4; 7 5; 7 2; 7 5; 7 5; 7 2; 7 5; 7 6; 7 5; 7 5; 7 6; 7 2; 7 6; 7 2; 7 6; 7 4; 7 4; 6 5; 6 5; 6 2; 6 4; 6 4; 6 2; 6 4; 6 2; 6 4; 6 5; 6 1; 6 4; 6 4; 6 4; 6 1; 6 2; 6 4; 6 3; 6 4; 6 3; 6 5; 6 5; 6 5; 6 4; 6 4; 6 4; 6 3; 6 4; 6 3; 6 5; 6 5; 6 2; 6 5; 6 4; 6 3; 6 2; 6 4; 6 4; 6 5; 6 4; 6 3; 6 4; 6 5; 6 2; 6 2; 6 4; 6 4; 6 3; 6 5; 6 4; 6 2; 6 3; 6 4; 6 3; 6 3; 6 3; 6 4; 6 4; 6 5; 6 5; 6 4; 6 4; 6 4; 6 5; 6 3; 6 2; 6 3; 6 2; 6 4; 6 4; 6 5; 6 2; 6 5; 6 4; 6 5; 6 3; 6 4; 6 4; 6 3; 6 3; 6 4; 6 1; 6 3; 6 3; 6 5; 6 5; 6 4; 6 5; 6 4; 6 4; 6 4; 6 3; 6 5; 6 5; 6 3; 6 4; 6 5; 6 5; 6 5; 6 5; 6 3; 6 3; 6 2; 6 2; 6 4; 6 5; 6 3; 6 4; 6 3; 6 4; 6 3; 6 5; 6 5; 6 2; 6 4; 6 2; 6 5; 6 4; 6 5; 6 4; 6 2; 6 3; 6 4; 6 5; 6 1; 6 3; 6 3; 6 3; 6 5; 6 3; 6 3; 6 1; 6 1; 6 2; 6 5; 6 5; 6 3; 6 3; 6 4; 6 4; 5 3; 5 3; 5 3; 5 4; 5 4; 5 2; 5 4; 5 4; 5 2; 5 4; 5 3; 5 2; 5 2; 5 4; 5 4; 5 3; 5 4; 5 4; 5 0; 5 4; 5 3; 5 3; 5 2; 5 4; 5 4; 5 4; 5 3; 5 3; 5 3; 5 1; 5 1; 5 4; 5 4; 5 4; 5 3; 5 3; 5 4; 5 3; 5 2; 5 3; 5 2; 5 4; 5 3; 5 2; 5 1; 5 2; 5 4; 5 3; 5 3; 5 3; 5 4; 5 3; 5 4; 5 4; 5 3; 5 1; 5 4; 5 4; 5 3; 5 4; 5 1; 5 4; 5 4; 5 3; 5 4; 5 3; 5 4; 5 3; 5 3; 5 4; 5 4; 5 4; 5 1; 5 4; 5 2; 5 4; 5 4; 5 2; 5 3; 5 3; 5 4; 5 4; 5 4; 5 3; 5 4; 5 2; 5 3; 5 4; 5 3; 5 3; 5 2; 5 3; 5 4; 5 3; 5 1; 5 1; 5 4; 5 3; 5 3; 5 3; 5 4; 5 4; 5 3; 5 4; 5 2; 5 3; 5 2; 5 4; 5 4; 5 4; 5 2; 5 2; 5 4; 5 1; 5 4; 5 0; 5 3; 5 4; 5 4; 5 2; 5 4; 5 3; 5 2; 5 2; 5 4; 5 2; 5 3; 5 3; 5 4; 5 2; 5 2; 5 4; 5 2; 5 2; 5 2; 5 2; 5 4; 5 2; 5 4; 5 4; 5 4; 5 4; 5 2; 5 4; 5 1; 5 3; 5 3; 5 4; 5 3; 5 2; 5 4; 5 4; 5 4; 5 4; 5 3; 5 2; 5 4; 5 2; 5 3; 5 3; 5 2; 5 2; 5 2; 5 3; 5 4; 5 4; 5 4; 5 4; 5 2; 5 4; 5 4; 5 3; 5 4; 5 4; 5 4; 5 3; 5 3; 5 4; 5 4; 5 4; 5 2; 5 3; 5 3; 5 4; 5 3; 5 3; 5 2; 5 0; 5 3; 5 3; 5 3; 5 4; 5 3; 5 3; 5 3; 5 2; 5 3; 5 4; 5 4; 5 4; 5 2; 5 1; 5 3; 5 2; 5 2; 5 1; 5 4; 5 4; 5 4; 5 4; 5 3; 5 3; 5 2; 5 4; 5 4; 5 3; 5 2; 5 4; 5 4; 5 4; 5 4; 5 1; 5 4; 5 4; 5 4; 5 1; 5 3; 5 2; 5 4; 5 4; 5 3; 5 4; 5 4; 5 4; 4 3; 4 2; 4 3; 4 3; 4 1; 4 3; 4 2; 4 3; 4 2; 4 2; 4 2; 4 3; 4 1; 4 3; 4 3; 4 2; 4 3; 4 1; 4 3; 4 2; 4 1; 4 2; 4 2; 4 0; 4 3; 4 3; 4 3; 4 2; 4 3; 4 2; 4 3; 4 2; 4 3; 4 3; 4 1; 4 2; 4 2; 4 3; 4 1; 4 3; 4 2; 4 3; 4 3; 4 3; 4 3; 4 3; 4 1; 4 2; 4 2; 4 3; 4 3; 4 2; 4 2; 4 3; 4 3; 4 3; 4 3; 4 2; 4 3; 4 2; 4 2; 4 1; 4 2; 4 3; 4 2; 4 1; 4 3; 4 3; 4 3; 4 3; 4 2; 4 2; 4 2; 4 2; 4 3; 4 3; 4 2; 4 1; 4 2; 4 3; 4 3; 4 2; 4 3; 4 3; 4 1; 4 3; 4 1; 4 2; 4 3; 4 3; 4 2; 4 0; 4 3; 4 1; 4 3; 4 2; 4 3; 4 2; 4 3; 4 2; 4 1; 4 2; 4 2; 4 3; 4 2; 4 3; 4 2; 4 3; 4 3; 4 2; 4 3; 4 1; 4 3; 4 1; 4 3; 4 2; 4 3; 4 3; 4 1; 4 2; 4 3; 4 3; 4 3; 4 2; 4 2; 4 1; 4 2; 4 3; 4 0; 4 3; 4 2; 4 2; 4 3; 4 3; 4 3; 4 2; 4 2; 4 0; 4 3; 4 2; 4 2; 4 2; 4 3; 4 2; 4 3; 4 3; 4 1; 4 1; 4 0; 4 3; 4 3; 4 3; 4 1; 4 2; 4 3; 4 3; 4 2; 4 1; 4 2; 4 2; 4 1; 4 3; 4 3; 4 1; 4 2; 4 2; 4 3; 4 1; 4 3; 4 2; 4 3; 4 3; 4 2; 4 3; 4 0; 4 2; 4 2; 4 3; 4 3; 4 2; 4 2; 4 3; 4 3; 4 3; 4 3; 4 3; 4 3; 4 3; 4 3; 4 3; 4 0; 4 3; 4 3; 4 3; 4 3; 4 1; 4 3; 4 1; 4 1; 4 1; 4 3; 4 2; 4 1; 4 2; 4 3; 4 3; 4 3; 4 3; 4 1; 4 2; 4 3; 4 3; 4 2; 4 3; 4 2; 4 3; 4 3; 4 1; 4 1; 4 3; 4 3; 4 3; 4 3; 4 2; 4 3; 4 3; 4 2; 4 2; 4 3; 4 2; 4 1; 4 3; 4 3; 4 2; 4 0; 4 2; 4 1; 4 2; 4 3; 4 1; 4 3; 4 3; 4 3; 4 3; 4 2; 4 1; 4 2; 4 3; 4 2; 4 3; 4 3; 4 2; 4 3; 4 3; 4 3; 4 3; 4 2; 4 3; 4 2; 4 3; 4 2; 4 2; 4 3; 4 2; 4 3; 4 0; 4 1; 4 3; 4 2; 4 3; 4 1; 4 3; 4 2; 4 2; 4 0; 4 3; 4 2; 4 3; 4 3; 4 3; 4 3; 4 2; 4 3; 4 3; 4 2; 4 2; 4 3; 4 1; 4 2; 4 2; 4 3; 4 2; 4 2; 4 3; 4 2; 4 3; 4 1; 4 2; 4 1; 4 2; 4 2; 4 2; 4 1; 4 3; 4 2; 4 3; 4 3; 4 3; 4 3; 4 2; 4 3; 4 3; 4 2; 4 3; 4 2; 4 3; 4 1; 4 3; 4 2; 4 3; 4 1; 4 3; 4 3; 4 2; 4 3; 4 3; 4 2; 4 2; 4 2; 4 3; 4 3; 4 2; 4 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 0; 3 1; 3 2; 3 1; 3 2; 3 1; 3 2; 3 1; 3 2; 3 0; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 0; 3 1; 3 1; 3 2; 3 1; 3 1; 3 2; 3 2; 3 0; 3 2; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 0; 3 2; 3 2; 3 2; 3 1; 3 0; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 2; 3 1; 3 1; 3 2; 3 1; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 0; 3 1; 3 2; 3 1; 3 1; 3 1; 3 2; 3 2; 3 2; 3 1; 3 1; 3 0; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 1; 3 1; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 0; 3 1; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 0; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 1; 3 0; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 1; 3 1; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 0; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 0; 3 2; 3 2; 3 1; 3 1; 3 1; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 0; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 0; 3 1; 3 1; 3 1; 3 2; 3 0; 3 2; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 1; 3 1; 3 2; 3 1; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 0; 3 0; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 0; 3 1; 3 1; 3 2; 3 2; 3 2; 3 0; 3 2; 3 2; 3 1; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 1; 3 0; 3 2; 3 2; 3 2; 3 0; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 0; 3 2; 3 1; 3 0; 3 1; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 1; 3 1; 3 1; 3 2; 3 1; 3 1; 3 2; 3 0; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 1; 3 1; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 1; 3 2; 3 2; 3 0; 3 0; 3 2; 3 1; 3 1; 3 2; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 1; 3 2; 3 2; 3 1; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 0; 3 2; 3 2; 3 0; 3 1; 3 2; 3 2; 3 2; 3 2; 3 0; 3 0; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 0; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 1; 3 1; 3 1; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 0; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 0; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 2; 3 1; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 1; 3 2; 3 1; 3 1; 3 0; 3 1; 3 2; 3 1; 3 2; 3 2; 3 2; 3 0; 3 1; 3 2; 3 1; 3 1; 3 2; 3 2; 3 1; 3 2; 3 2; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 0; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 0; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 0; 2 0; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 1; 2 0; 2 1; 2 0; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 0; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 0; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; 2 1; "
     ]
    }
   ],
   "source": [
    "for i, group in enumerate(smart_modified_HOGs):\n",
    "    selected_rows = counts['Desc'].isin(group)\n",
    "    if len(group) != selected_rows.sum():\n",
    "        print(len(group), selected_rows.sum(), end='; ')\n",
    "    new_row = counts[selected_rows].sum(numeric_only=True)\n",
    "    badge_n = i + OG0064650\n",
    "    badge = 'JSG' + str(badge_n).zfill(7)\n",
    "    new_row['Desc'] = badge\n",
    "    new_row['Family ID'] = badge_n\n",
    "    new_table.loc[badge] = new_row\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17044\n"
     ]
    }
   ],
   "source": [
    "from itertools import chain\n",
    "rows_mentioned = set(chain(*smart_modified_HOGs))\n",
    "print(len(rows_mentioned))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['OG0018821',\n",
       " 'OG0022545',\n",
       " 'OG0026509',\n",
       " 'OG0036072',\n",
       " 'OG0026401',\n",
       " 'OG0017896',\n",
       " 'OG0009334',\n",
       " 'OG0006179',\n",
       " 'OG0029711',\n",
       " 'OG0031620']"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(rows_mentioned)[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How could there possibly be numbers mentioned here that are not in the table?  Can I use a mask instead?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21603\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Desc</th>\n",
       "      <th>Family ID</th>\n",
       "      <th>FRAX00</th>\n",
       "      <th>FRAX01</th>\n",
       "      <th>FRAX02</th>\n",
       "      <th>FRAX03</th>\n",
       "      <th>FRAX04</th>\n",
       "      <th>FRAX05</th>\n",
       "      <th>FRAX06</th>\n",
       "      <th>FRAX07</th>\n",
       "      <th>...</th>\n",
       "      <th>FRAX28</th>\n",
       "      <th>FRAX29</th>\n",
       "      <th>FRAX30</th>\n",
       "      <th>FRAX31</th>\n",
       "      <th>FRAX32</th>\n",
       "      <th>FRAX33</th>\n",
       "      <th>FRAX34</th>\n",
       "      <th>Mguttatus</th>\n",
       "      <th>Oeuropea</th>\n",
       "      <th>Slycopersicum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OG0000000</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>22</td>\n",
       "      <td>29</td>\n",
       "      <td>17</td>\n",
       "      <td>35</td>\n",
       "      <td>23</td>\n",
       "      <td>28</td>\n",
       "      <td>23</td>\n",
       "      <td>...</td>\n",
       "      <td>27</td>\n",
       "      <td>27</td>\n",
       "      <td>22</td>\n",
       "      <td>28</td>\n",
       "      <td>24</td>\n",
       "      <td>23</td>\n",
       "      <td>20</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OG0000001</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>20</td>\n",
       "      <td>15</td>\n",
       "      <td>19</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>18</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>9</td>\n",
       "      <td>20</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OG0000002</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>15</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>17</td>\n",
       "      <td>18</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "      <td>9</td>\n",
       "      <td>22</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OG0000003</td>\n",
       "      <td>3</td>\n",
       "      <td>24</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>34</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OG0000004</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>16</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "      <td>16</td>\n",
       "      <td>15</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>14</td>\n",
       "      <td>15</td>\n",
       "      <td>15</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Desc  Family ID  FRAX00  FRAX01  FRAX02  FRAX03  FRAX04  FRAX05  \\\n",
       "0  OG0000000          0      11      22      29      17      35      23   \n",
       "1  OG0000001          1      22      20      15      19      20      16   \n",
       "2  OG0000002          2      19      15      12      15      16      17   \n",
       "3  OG0000003          3      24      10       6      12      15      13   \n",
       "4  OG0000004          4      20      16      11      10      13      15   \n",
       "\n",
       "   FRAX06  FRAX07      ...        FRAX28  FRAX29  FRAX30  FRAX31  FRAX32  \\\n",
       "0      28      23      ...            27      27      22      28      24   \n",
       "1      20      17      ...            16      16      20      20      18   \n",
       "2      18      17      ...            17      15      17      15      13   \n",
       "3      15      17      ...            13      19      12      15      14   \n",
       "4      16      15      ...            13      14      15      15      14   \n",
       "\n",
       "   FRAX33  FRAX34  Mguttatus  Oeuropea  Slycopersicum  \n",
       "0      23      20          1        19              5  \n",
       "1      20      20          9        20             10  \n",
       "2      18      18          9        22             13  \n",
       "3      13      12         21        34             24  \n",
       "4      15      15          8        13             12  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "old_rows_removed = new_table[~new_table['Desc'].isin(rows_mentioned)]\n",
    "print(len(old_rows_removed))\n",
    "old_rows_removed.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.int64"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(new_table.loc[0]['Family ID']) #counts['Desc'] == 'OG0000000'][0]#.type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33693\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Desc</th>\n",
       "      <th>Family ID</th>\n",
       "      <th>FRAX00</th>\n",
       "      <th>FRAX01</th>\n",
       "      <th>FRAX02</th>\n",
       "      <th>FRAX03</th>\n",
       "      <th>FRAX04</th>\n",
       "      <th>FRAX05</th>\n",
       "      <th>FRAX06</th>\n",
       "      <th>FRAX07</th>\n",
       "      <th>...</th>\n",
       "      <th>FRAX28</th>\n",
       "      <th>FRAX29</th>\n",
       "      <th>FRAX30</th>\n",
       "      <th>FRAX31</th>\n",
       "      <th>FRAX32</th>\n",
       "      <th>FRAX33</th>\n",
       "      <th>FRAX34</th>\n",
       "      <th>Mguttatus</th>\n",
       "      <th>Oeuropea</th>\n",
       "      <th>Slycopersicum</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>JSG0033688</th>\n",
       "      <td>JSG0033688</td>\n",
       "      <td>33688</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JSG0033689</th>\n",
       "      <td>JSG0033689</td>\n",
       "      <td>33689</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JSG0033690</th>\n",
       "      <td>JSG0033690</td>\n",
       "      <td>33690</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JSG0033691</th>\n",
       "      <td>JSG0033691</td>\n",
       "      <td>33691</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JSG0033692</th>\n",
       "      <td>JSG0033692</td>\n",
       "      <td>33692</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Desc  Family ID  FRAX00  FRAX01  FRAX02  FRAX03  FRAX04  \\\n",
       "JSG0033688  JSG0033688      33688       3       3       3       3       3   \n",
       "JSG0033689  JSG0033689      33689       2       2       1       2       2   \n",
       "JSG0033690  JSG0033690      33690       2       1       2       1       1   \n",
       "JSG0033691  JSG0033691      33691       2       2       0       1       1   \n",
       "JSG0033692  JSG0033692      33692       1       0       1       0       0   \n",
       "\n",
       "            FRAX05  FRAX06  FRAX07      ...        FRAX28  FRAX29  FRAX30  \\\n",
       "JSG0033688       3       2       3      ...             3       3       3   \n",
       "JSG0033689       2       2       1      ...             2       0       0   \n",
       "JSG0033690       1       1       1      ...             0       1       1   \n",
       "JSG0033691       2       2       1      ...             1       1       2   \n",
       "JSG0033692       0       0       0      ...             1       2       2   \n",
       "\n",
       "            FRAX31  FRAX32  FRAX33  FRAX34  Mguttatus  Oeuropea  Slycopersicum  \n",
       "JSG0033688       2       3       2       3          2         3              3  \n",
       "JSG0033689       2       2       2       2          1         2              1  \n",
       "JSG0033690       1       1       1       1          1         1              1  \n",
       "JSG0033691       2       1       1       1          1         2              1  \n",
       "JSG0033692       0       1       1       0          0         0              0  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(new_table))\n",
    "new_table.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Up until now we haven't ran a clade filter because combining rows might join some things and make it pass the filter.  Now we need to filter out our rows so that we have usable families for CAFE.  There should be at least one copy in an Oleaceae species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## New Plan: only use OrthoGroups.csv\n",
    "Cafe_orthofinder_Orthogroups.GeneCount.csv is filtered to have at least two genes in any family.  We don't want this filter apriori, only after clustering.  \n",
    "\n",
    "- Important: super HOGs network need to be redone with this as its new source.\n",
    "- Important: start pulling of the _R1 tail, that's not helpful."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Old code using FASTA sequence files for HOG definitions, useful for MSA viz output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def parse_HOGs_from_fasta_directory():\n",
    "    directory_HOGs = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\Ash_Proteome\\Results_Jun25\\Orthologues_Jul06\\Sequences\"\n",
    "    from glob import glob\n",
    "    from os.path import join, splitext, basename\n",
    "    file_list = glob(join(directory_HOGs, \"*.fa\"))\n",
    "    HOGs = {}\n",
    "    for i, filename in enumerate(file_list):\n",
    "        if i % 1000 == 0:\n",
    "            print(basename(filename))\n",
    "        assert basename(filename).startswith('OG')\n",
    "        headers = []\n",
    "        with open(filename, 'r') as fasta:\n",
    "            for line in fasta:\n",
    "                if line.startswith('>'):\n",
    "                    headers.append(convert_to_gene_id(line[1:-1]))\n",
    "        #assert len(headers) == len(set(headers)), \"There was a redundant gene mention %s\" % headers \n",
    "        HOGs[basename(splitext(filename)[0])] = headers\n",
    "    return HOGs\n",
    "HOGs = parse_HOGs_from_fasta_directory()\n",
    "len(HOGs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_We could simply exclude all modified HOGs from the data as being ambiguous_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concat super HOG fasta files into a new directoy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from os.path import join, basename, splitext\n",
    "import os\n",
    "family_out_dir = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\Ash_Proteome\\Results_Jun25\\Orthologues_Jul06\\super_JOGs__tail\"\n",
    "HOGfasta = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\Ash_Proteome\\Results_Jun25\\Orthologues_Jul06\\Sequences\"\n",
    "os.makedirs(family_out_dir)\n",
    "for family in modified_HOGs_only:\n",
    "    family_name = ('-'.join(family))[:200] + '.fa'\n",
    "    with open(join(family_out_dir, family_name), 'wb') as concat:  # output file concatenation\n",
    "        for hog in family:\n",
    "            with open(join(HOGfasta, hog + '.fa'), 'rb') as hog_file:  # smaller input file\n",
    "                concat.write(hog_file.read())  # dump the whole file\n",
    "#                 concat.write('\\n')  # just to be safe\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Align the new gene family files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import multiprocessing\n",
    "from datetime import datetime\n",
    "import os\n",
    "from os.path import dirname, join, basename\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from DNASkittleUtils.CommandLineUtils import just_the_name\n",
    "from glob import glob\n",
    "\n",
    "def do_alignment(args):\n",
    "#     print(\"starting alignment\")\n",
    "    index, fa, output_folder = args\n",
    "    target = join(output_folder, just_the_name(fa) + '.fa')\n",
    "    muscle_exe = 'muscle3.8.31_i86win32.exe'\n",
    "\n",
    "    if not os.path.exists(target):\n",
    "        muscle_cline = MuscleCommandline(muscle_exe, input=fa, out=target)\n",
    "        try:\n",
    "            stdout, stderr = muscle_cline()\n",
    "        except subprocess.CalledProcessError as err:\n",
    "            print(err.stderr)\n",
    "        print(datetime.now(), just_the_name(fa), '{:%}'.format(index / 5347))\n",
    "\n",
    "def batch_align_sequences(input_folder, output_folder):\n",
    "    start = datetime.now()\n",
    "    input_folder = os.path.abspath(input_folder)\n",
    "    os.makedirs(output_folder, exist_ok=True)\n",
    "    files = glob(os.path.join(input_folder, '*.fa'))\n",
    "    files = sorted(files, key= os.path.getsize)\n",
    "    args = [(i, ipath, output_folder) for i, ipath in list(enumerate(files))]\n",
    "    for line in args:\n",
    "        do_alignment(line)\n",
    "#     pool.map(do_alignment, args)\n",
    "        \n",
    "    return os.path.abspath(output_folder)\n",
    "\n",
    "#You can't actually do multiprocessing from a notebook\n",
    "# if __name__ == '__main__':  # https://github.com/jupyter/notebook/issues/2080\n",
    "#     pool = multiprocessing.Pool(10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "family_aligned_dir = r\"D:\\Genomes\\Ash_Gene_Families\\test\\aligned\"   \n",
    "family_fasta_dir = r\"D:\\Genomes\\Ash_Gene_Families\\test\"\n",
    "\n",
    "batch_align_sequences(input_folder = family_fasta_dir,\n",
    "                      output_folder = family_aligned_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "family_aligned_dir = r\"D:\\Genomes\\Ash_Gene_Families\\aligned\"   \n",
    "family_fasta_dir = r\"D:\\Genomes\\Ash_Gene_Families\"\n",
    "\n",
    "batch_align_sequences(input_folder = family_fasta_dir,\n",
    "                      output_folder = family_aligned_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "family_aligned_dir = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\data\\super_hog_test_viz\\aligned\"   \n",
    "family_fasta_dir = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\data\\super_hog_test_viz\"\n",
    "\n",
    "batch_align_sequences(input_folder = family_fasta_dir,\n",
    "                      output_folder = family_aligned_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "family_aligned_dir = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\data\\phylogenetic_guides\\aligned\"   \n",
    "family_fasta_dir = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\data\\phylogenetic_guides\"\n",
    "\n",
    "batch_align_sequences(input_folder = family_fasta_dir,\n",
    "                      output_folder = family_aligned_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "do_alignment((1, r\"D:\\Genomes\\Ash_Gene_Families\\OG0037028-OG0004348.fa\",\n",
    "             r\"D:\\Genomes\\Ash_Gene_Families\\test\\aligned\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "do_alignment((1, r\"D:\\Genomes\\Ash_Gene_Families\\test\\FRAX04_pair.fa\", r'D:\\Genomes\\Ash_Gene_Families\\test\\aligned'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "list(enumerate(glob(join(r'D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\data\\candidates', '*.fa'))))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* output files with lists of gene names to look for in each HOG\n",
    "* per species, find those gene names in the annotation\n",
    "* count presence / absence of a gene name in an annotation\n",
    "* group them back by super-HOGs\n",
    "* End result: gene copy count per each gene family defined by a super-HOG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_Should I call this a multigene family?_"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fasta Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from DNASkittleUtils.Contigs import read_contigs\n",
    "seqs = read_contigs(r\"D:\\Genomes\\Ash_Gene_Families\\OG0014503-OG0018028-OG0009430-OG0001620-OG0012608-OG0012187-OG0003830-OG0005353-OG0000848-OG0001316-OG0001679-OG00.fa\")\n",
    "print(len(seqs))\n",
    "%matplotlib inline\n",
    "import pandas\n",
    "lengths = [len(contig.seq) for contig in seqs ]\n",
    "df = pandas.Series(lengths)#, columns=['Family ID', 'Copies'])\n",
    "bins = 20\n",
    "f = df.hist(bins=bins, figsize=(10,5), )#dpi=300)\n",
    "f.grid(False)\n",
    "# f.set_xticks(range(bins))\n",
    "f.set_title(\"Are sequences clustered together actually the same size?\")\n",
    "f.set_xlabel(\"Sequence Length\")\n",
    "f.set_ylabel(\"Number of Sequences\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Conclusion: This indicates that the file I opened is a massive super cluster with over 1,000 sequences in the 500 range and many more besides.  Making a cluster this big is not helpful to my analysis.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for family in modified_HOGs_only:\n",
    "    if 'OG0018028' in family:\n",
    "        print(family)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "len({'OG0023682', 'OG0001620', 'OG0008375', 'OG0036452', 'OG0019438', 'OG0000848', 'OG0025173', 'OG0007534', 'OG0001316', 'OG0019062', 'OG0002719', 'OG0012608', 'OG0003830', 'OG0010137', 'OG0023528', 'OG0007285', 'OG0014503', 'OG0011854', 'OG0010197', 'OG0020872', 'OG0001331', 'OG0009430', 'OG0018028', 'OG0010551', 'OG0031433', 'OG0031546', 'OG0001679', 'OG0012187', 'OG0005353'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def i_hist(values):\n",
    "    bins = [0] * (max(values) +1)\n",
    "    for v in values:\n",
    "        bins[v] += 1\n",
    "    return bins\n",
    "i_hist([len(family) for family in modified_HOGs_only])  # cap at 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "list(zip(range(400), i_hist([len(genes) for genes in HOGs.values()]) ))  # cap at 200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "i_hist([len(ogs) for ogs in actual_genes_to_HOGs.values()])  # cap at 6 HOGs per gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Blacklist promiscuous genes and HOGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.text.Text at 0x2edb2b38>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA2oAAAHUCAYAAABLdNsnAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAAPYQAAD2EBqD+naQAAIABJREFUeJzs3XeYZGWd9//3B0QQVxCdJbgrZhRdcWUMqGBCYVVMGEf4\nmR8zrqhreAwgrgkDBnR1FQOi4/ogZhSFVYKoKIOKMowJxDTIGAZF8nx/f9ynoKamp6e7p3r6NP1+\nXVddQ51zn1PfU6GpT933uU+qCkmSJElSf2w21wVIkiRJktZmUJMkSZKknjGoSZIkSVLPGNQkSZIk\nqWcMapIkSZLUMwY1SZIkSeoZg5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJM0q5IcmmTNFNuuSfLa\n2a5pXJLcoqv5yXNdy7hM5/WSktyv+wzcd0z78/0nSR2DmqQpS/KU7kvZ4HZpkt8m+WqSg5L8wwSb\nFeAXr/mjuptmUZJ7JTkkyTZzXcsYjPP9Mq/ff0nOT/KF9awbhNr9J1h3xyTHJPlNksu6v6vHJLnj\nJI91yyRHJlmR5JLu9pNu2Z2nWXeSPDHJ/yb5Q5KLknwmyS2msx9J42VQkzRdBbwaOBB4DvDubtk7\ngbMn+ILwemDrTVqh1H/3Bl4L3HiuC9FYbShkrrO+C27LgAcAHwaeC3wIuD+wLMkjJ9hmP+DHwAHA\n14EXAS8EjgceApyV5ObTqHtn4GPAr4H/C7wH2Bv4SpItprEfSWN0vbkuQNK89NWqWjZ0/y1J7g98\nGfh8kl2r6nKAqloDXDEHNUp9lik3TAJcf/CZ0tQl2aqqLpvrOoas9bonuTVwNPBz4L5V9aehde8C\nTgM+nmS3qjp/aJulwHnA3lX1h5F9vgx4PtMbyfBn4M5V9dOh/awE/gu4G/DtaexL0pjYoyZpLKrq\nm7Tes1vQetuAic85SXL9JEd0Q2wuTvK5JP80lcfphuS8beh+kvwlyZXDw8iSvLxbtvXQstsnOTbJ\nH7thm99L8vApPu62ST7aPdafk3yECXpDktw5yUeS/KJ7jN8nOSrJTUbaHdoNg7pdN8TpL93zcVi3\n/ubd87K628eLR7YfDKN6fJI3dm3+luTzSf55ise0Z/ccXJrkZ0metZ52myd5TZKfd8Oyzkvyn0mu\nP9Tm7UkuGtnuPV2NLxhatn237Fkjx/G4JK9K8uuunhOT3GaKx3HXJF/pnqu/dtvec6TNYNjuvZO8\no3uu/5bkuCQ3neLjbPD9MzR0bNHQsi2SnN09xzdIcghweLf6/K6uq5Ps3LVfk+TdSZ6U5MfAZcC+\n3bokeVGSH3c1rEzy/iQ3Hqnj/CRf6J7f7yX5e5IfJblft37/7v6lSb6f5F9ncryDpkPbvC7JFRM9\np0n+O8mfht83U5HkaUlOSnJh9/77SZLnTNBucMz7dLVeBgzeZ1t1z+lFufZvzs0ywXmx3fIPd8/t\nZd1z/fTp1DwNLwNuADxrOKQBdPefA/xD127g5bRRCk8bDWnddlVVR1bVbwfLkuyQ9nfp190x/a57\nDnbutrl4OKR1BgF3Wq+XpPExqEkap4/TvrTtM7RsonNOjqIN0/kq7UvHlbTeuKmcm/ItYHjigt2A\nQUC7z9DyPYFlVfV3gCR3Ar4D3B54E/Bi4G/A5zLB0KIJfIE2zOho4FXAP9OGCo3W/GDgVrQhTC+g\n/fL9xO74hg22+5/u35d39b0qyYuArwG/6Zb/DHhrkj0nqOtVtKFObwbe1T3+15NsOdnBJPkX4ARg\nEW0I3keAQ4FHT9D8KOB1wPdpQ6y+SRsetXSozanATbL2OTV7AlcDew0tu2937KeMPMYrgEcCbwXe\nCOwBHDPZMXTHccduX3emPQeHAbcEvpnk7hNs8p6u7aHA+4CHA0dO4XGm+v55GrAV8P6hZYcBuwJP\nrapLgc9w7XP377QfNv4/YDjo7g28HfhU1+b8bvl/A2+hPd8vpL3PDgC+mmTzoe0LuB3wCdp79xXA\ndsAXkjyp2/fRtNf+Nlz7Ppzu8Q4ea+Bo2midJ4zsbwvgMcCxVTXdHvbn0I7/DV0dFwDvS/LcCeq4\nA/BJ2ufnIOAH3bqP0XqZvkQLPZcywd+cJNsD3wUeSBvW/ULa5+9DSV44xXq3SHLT0RsTD3PdDzi/\nqk6faEdVdUp37PsNLX4Y8POq+v4U6wE4jvb5Ooo2tPJdtAC480SN035YehWwgtarJ2kuVJU3b968\nTekGPIX2xXv3Sdr8Gfj+0P1DgKuH7u9GG5Lz7pHtjun2/doN1PAS2lDKG3b3XwD8kjY0541D7f4E\nvG3o/onAWcD1RvZ3GnDuBh7zkV3NLx5aFuDkruYnDy3fcoLtn9C1u8/I87IGeN/Qss1oX0KvAl4y\ntHxb4BLgw0PL7tdtfwGw9dDyx3bLX7CBY/pst89/Glp2e1ponuj1ev/I9od3x3S/7v6irt2zu/vb\ndMfxKeB3Q9u9E7hoguP4MbD50PKDuv3fcQrHcSlwi6FlOwKrgW+MvHfX0IbtDm//9u79dKMNPM6U\n3z/A/+keawlwz+45fdtIm5d0x7fzBI+1ptvm9iPL9+zWPWFk+YO75U8cWnZet/97TtDub8A/j9R7\nNW3o3bSOt3v9Rrf9FnD6yHaP7trttYHnea2/F5N8pr4C/Gxk2eCYHzSy/K7dcY++Bh9m5G8O7dyw\n3wA3Hmn7SdrflHVqmaCGNZPcrgb2H/qMrAGO28A+P9dtd0PgRt02n5mg3bbATYduWw0tX+vv1wYe\n74a0H2UuGn0PevPmbdPe7FGTNG5/o32ZWJ+H0n7Ffs/I8ncytfN2TqX9Yn/v7v5e3bJTu/8myW60\nX69P7e5vRztR//8B2478yv014HZJdprkMR9C++J8TS9JVQ2OYa2aa+g8oiRbdo/x3a7d7iP7Ldov\n3INt19C+IIXWwzVYvpr2y/atJ6jtY9X1GnZtjwV+T3ueJ5RkM9qX9s/V0PCoqlpB62UbNni9jhhZ\n/vauzod1264CzuXa3s49aUHtrcCOuXYY415M/Av9h6vq6qH7p3b7n+iYR4/js1X1q6HjWEn7Yr1X\n1p6JtGg9UsNOBTanDdld3+NM6/1TVR+kBYkjaT1MP6P1TkzHN7vXY9hjgb8AJ43UcBbtc/eAkfbn\nVNV3h+4P/vukqvrNyPJrnusxfF6OBu6Zdi7VwAHAr6vq1A0c9zpGPlPbdHWcAtw6yejfmvOq6sSR\nZf9Ge+3/a2T5Op9fYH/gi8DmExz3tqz7GZ7Id2g9og8aub10pN2g9r9uYH+D9dtw7eiBv03Q7pu0\ncDW4Pa9bfintx4j7jw6RXY/3A3cEHjbBe1DSJuRkIpLG7R+ACydZfwvar7u/GFk+1S8Ey4C/077w\nf50WCA7pHvOg7vyXvWhfzAaB4La0L2SvB/5zgn0WsD0t4Kyv5t8PB6L11dx9yT2U1ou2/chjbDvB\nvi8Yub8auKxGzlfplt+Edf18Pcsmm1b7H2nnuPxsgnUraMF0YPB6rfU4VXVhkr+MPM5ptC/F0F6X\n71fVmUn+SAtNf6D10H1igsf99cj9P3f/bjeF4xg9twZgOe01v3n33xvzODN5//wf2nv8tsC9a/oT\ngZw/wbLb0X6AWOe8pKEahq313qqqi5NA6zEatrr7d/AcbOzn5X9oP7w8CfjPtHNHHwq8Yz3tJ5Xk\nPrSht3uw9gyyg8/UcNA5b4JdDN7Do+vWek8n+Ufa8/ss4NkT7Gei53giq6rqG6MLk1zN2sFwUPdk\nP2wNr/8r156yMtGlUJ7Vtd2BoWHDVXVFkpcDbwMuTPId2hDQo6tqrb/VSa4HPB44qqrO2EBdkmaZ\nQU3S2KRNCLItE4eHa5ptzGNU1VVJvgvct+ul2Yn26/pFwBa0oWZ7Asur6o/dZoMvN29j3R6jgQ3V\nPNH5cxMdy/+jfaE8HPgh7ZfvzbrHnWgUw9VTXLa+x5tJu8H6qRzTZG1HnQo8I8mtuLanE9pQuL1o\nX+w3Z93z02BmxzyT99JMHmcm758HAFvSnrc7c21v1lRdup46LqQFoInqvWjk/vqOdUPPwUZ9Xqrq\nL0m+ROtF+0/gcbTnYqKAPqmuV+5EWtg+mBa0r6D15L6IdT9TEz1v6y115P5gX8fQzmmbyI+msf/J\nH7wF59/TfryYzG7Ab6vqbwDdNv8ywf6+162/Bev29L8r7fpuj6JNTHMY8MokD6iqHw413Yb2d3Tl\nzI5K0jgZ1CSN05NpX36+Okmb82lfiG7D2j06d5jG45xKmxDgwbTznX4KkOQntKF3e9GGLw38svv3\nyqr632k8znDND0iy9Uiv2u2HG3XDih4IvKaq3jC0/LYzeMyput0Ey25DC4nr8wfaF9pdJlh3+5H7\n59Ner9sx1IPYTbpwY+BXQ20HwezBtCm939TdP4U2gcHvaefFDV/aYWP8gda7OloztMk7inV70GZi\nWu+fbljgu2kh5wrg7UlOqKrhWmZyUedf0IbUnT6DHrrp2NjPC7Thj59LcjdasDyrqpZvYJuJPJw2\n6+DDa+1ZDPeexj5+RXsP34q1e/JH3/8X0XqtNt+I456uLwHPTHLvmmBCkSR70SbHGR62+WXaDyJ3\nq2lMKFJV59GGMB/R/cj1Q9q5kk8eanYZbciuvWlSD3iOmqSxSPJA2oWwf0k7P2h9vkL7tXd0BrUX\nMfUvr6fSZtb7d9Y+3+k02ux5O3FtaKCqLqKdv/HsJDtOUPui0WUjjqf9ynzNLHPd+VEHjdQ86KkY\n/dt6MDP7Yj4VTx4+DyvJ42jHf/z6NujOhTsBeFSGpvJPsitrz9hJt5/QXp9hL6Ed0zWzWVa7ztPv\naMe7Ba0nDdprcRvaOVbf7h5/o3X7+RrwyME0491x7ECbyOOUQS/ERj7OdN8//017zp5OG0J3JUPn\nInYu6f6dzgWvP037gfW1oyvSLqEw0dDaaRvD5wXa5/yPtFlL70ebEXYm1vlMdcf51Gns4wTa6/G8\nkeVrfX6799NngMd0s16uZYrHPV1vpYWjD2TdS3jchHa+2CW03s2Bw2k/tHy4+8Fk1Fp/f9IuCTE6\nC+x5tFA6uvwK4L3A96Z5HJJmgT1qkqYrwEO7L/XXo50P8UBaL8p5wCNqkum3q+qHSZYCz+t6oE6n\n9RLchqkPZfs2baKKXYAPDC0f9NwUQ0Gt8/xu2dlJPkgLlDsA9wL+iTYz3Pp8kRYC39wN6zuHNunA\nWueWVNVfk5wCvKw7V+63tOBzq2kc23T9CTgt7bpuO9LC609ps9dN5hDa+WSnJXkfLVi9APgJbage\nAFX1oyQfA57VnX93Mm146ZNps9WdPLLfU2mXI/hRNwkKtB60S7h2uvhxejVtooZvdcdxNe1cneuz\n9rWnYP2vwVRemym9f5I8jTYs78lV9ftu2QuBY5I8t6oGPSNndo/7xiSfooW5L1Sbvn9CVXVKkg8A\nr0i77tnXuu12oYXgF9KmYR+H6Xxe1nn+uiHKn6K9pwazf87E4Bi/1B37jYBn0oaArhMiJ1JVy5J8\nBnhRF7a+QwuPg97o4R9RXgHcH/hud9zn0M4NXUz7OzfWsFZVP0/yFNpwy7OTHEX7O3orWtC/KW02\nz/NGtnkS7QexFUk+QesdS7fdk2ifg8G5iLvQJqD5dHc8V9H+fm3P2pfYgPbaLqedZ3vYOI9V0vQZ\n1CRNV9FO7If26+ufgLNpXxI/WlWXrGebYU+jDVs7gDb1/Um0L7e/nqDtujur+nuSs2hfnoZ71E7t\ntr9gZJgZVbW8G4Z1CG2q9pt2NZw1dDzre7xK8gjaBAkHdI/xedo1nc4aab6ENpvc82hfnE6gBaLf\nTeXYBg85xeVFu+bYbrQvmDeiTbDy/Kq6jElU1dlJ9qFN8PA62pe61wI3YyiodZ5BGzL2VNo5Litp\n17Sa6IvcqbSJVIZ7NK9O8m3aF92JZv2b6vFOdBzndMPD3kR7DjajfRF/0gTDwjbmcTb4/unO0XwH\n8PmqGp7M4ZNJHgO8JcnxVfWrqvp+klfTrhG2L9cOzbuAia89ONjXc5N8n9ZT9wbal+7zaUMNvzXc\ndD37mNLyaX5e1vf8HU0LaieOTlqxAcN1/LR77v6T1vu0knb9uz+ybi/lep83Wk/772mfz0fRPidP\noP2occ1npar+kOQetM/Co2k//PyR9gPGaPBfX+2TvZ/WWVdVxyZZDrySFs4WdY/5v8CbquqcCbb5\nQpI703q2H0z7m1q0YZ5fBD5QVWd3zX9NC3V7067ZdxVthtbHVdXnZnAMkjaRtBmmJUnzSZL7Ad8A\nHltV4+pFkcYm7TIZPwAOrKrJhkPPia5XchlwQFWN9ixJ0pyb83PUkjwnyQ+TrO5upyf5t6H1WyZ5\nb5JVSf6a5NjRMdlJbp7ky0kuSbIyyeHd+SPDbe6f5MwklyX5aTfUQJIkzY5n0c6D+uxcFzLBOVrQ\nzru8molnIZWkOdeHoY+/pp1sPJjq96nA55P8azdD1Dtp1/R5DHAx7STXz3DthW03o53s/jvalNg3\no520fAXt3AWS3JI2s9L7aGO3HwR8KMnvqurrs32AkiQtFEn2A+5Eu5bcuyc7724TelmSxbRJUq6i\nXddtX9oQwd9OtqEkzZVeDn1MuzjqS2mB7CLaibSf7dbdnnai6x5VdUaShwBfAHaqqlVdm2cDbwb+\nsTuh+S3AQ6pqt6HHWApsW1UP3ZTHJknj0A19/F/aeSYOfVRvJDmPNlHFV2mTqkx03uomleRBtPPO\n7ki7WPQFtHPo3jiuWUgladz60KN2ja537PHA1rRZ3RbTajxp0KaqViS5gDbz1Bm0XrSzByGtcwLt\nmiN3os2EtAftgpmMtDlido5EkmZXN9vi5nNdhzSqqm411zWMqqoTWfd7gCT1Wi+CWpJ/oQWzrWjj\n2R9dVecmuStwRVVdPLLJ8LS8O3b3R9cP1v1wkjbbJNlyfRcOTXJT2tCI8xmaFUqSJEnSgrMV7SL0\nJ1TVH2f7wXoR1GjTxN6FduHPxwBHJ7nvJO3D1KaOnaxNptBmX8Z/zR9JkiRJ89cBtMtezKpeBLWq\nuop2MU2AZd01TP4d+DRw/STbjPSqbc+1PWQrgbuP7HKHoXWDf3cYabM9cPFkF+al9aRxzDHHsOuu\nu07xaNQHBx98MEcc4cjW+cjXbv7ytZuffN3mL1+7+cvXbn5avnw5Bx54IHQZYbb1IqhNYDNgS+BM\n2uxMe9NN75tkF2Bn4PSu7beB/5tk0dB5avsAq2mTjgzaPGTkMfbplk/mMoBdd92V3XfffcYHo01v\n22239TWbp3zt5i9fu/nJ123+8rWbv3zt5r1NckrUnAe1JG8AvkKbpv9GtK7E+wH7VNXFSY4C3pHk\nz7Tz194NfKuqvtft4mvAOcDHk7wc2Al4PXBkVV3ZtXk/8IJu9scP04LfY2nT80qSJElSr8x5UKMN\nSTyaFrBWAz+ihbT/7dYfTLsg5bG0XravAs8fbFxVa7prtvwXrZftEuCjwCFDbc5P8jDgHcALgd8A\nz+hmgZIkSZKkXpnzoFZVz9zA+suBg7rb+tr8GthvA/s5mTbdvyRJkiT12mZzXYA0G5YsWTLXJWiG\nfO3mL1+7+cnXbf7ytZu/fO00Famayiz3C1OS3YEzzzzzTE/4lCRJkhawZcuWsXjxYoDFVbVsth/P\nHjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrGoCZJkiRJPWNQkyRJkqSeMahJkiRJUs8Y1CRJkiSpZwxq\nkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrGoCZJkiRJPWNQkyRJkqSeMahJkiRJUs8Y1CRJ\nkiSpZwxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrGoCZJkiRJPWNQkyRJkqSeMahJkiRJ\nUs8Y1CRJkiSpZwxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrGoCZJkiRJPWNQkyRJkqSe\nMahJkiRJUs8Y1CRJkiSpZwxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrGoCZJkiRJPWNQ\nkyRJkqSeMahJkiRJUs8Y1CRJkiSpZwxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1zPXmuoD54PGPfxI3\nuMHWc13GNV7/+tfyqEc9aq7LkCRJkjRLDGpT8ItfbAXsMddldL7Mxz72MYOaJEmSdB1mUJuSfYDD\n57qIzn5zXYAkSZKkWeY5apIkSZLUMwY1SZIkSeoZg5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJMk\nSZKknjGoSZIkSVLPzHlQS/LKJGckuTjJhUk+m2SXkTbfTLJm6HZ1kveNtLl5ki8nuSTJyiSHJ9ls\npM39k5yZ5LIkP03ylE1xjJIkSZI0HXMe1IC9gPcA9wQeBGwBfC3JDYbaFPDfwA7AjsBOwMsGK7tA\ndjztAt57AE8BngocNtTmlsCXgJOAuwDvAj6U5MGzclSSJEmSNEPXm+sCquqhw/eTPBX4A7AYOG1o\n1d+r6qL17GZf4A7AA6pqFXB2ktcAb05yaFVdBTwX+GVVDQLeiiR7AgcDXx/bAUmSJEnSRupDj9qo\nG9N60P40svyAJBclOTvJG0d63PYAzu5C2sAJwLbAnYbanDiyzxOAe42vdEmSJEnaeHPeozYsSYB3\nAqdV1TlDqz4B/Ar4HbAbcDiwC/DYbv2OwIUju7twaN0PJ2mzTZItq+rycR2HJEmSJG2MXgU14H3A\nHYH7DC+sqg8N3f1JkpXASUluVVXnbWCfNcm6TKENcBxw7siyJd1NkiRJ0nXJ0qVLWbp06VrLVq9e\nvUlr6E1QS3Ik8FBgr6r6/Qaaf7f797bAecBK4O4jbXbo/l059O8OI222By6uqismf7j9aZ14kiRJ\nkq7rlixZwpIla3fKLFu2jMWLF2+yGnpxjloX0h5JmwzkgilscldaL9gg0H0buHOSRUNt9gFWA8uH\n2uw9sp99uuWSJEmS1BtzHtS666EdADwJuCTJDt1tq279rZO8OsnuSW6R5BHAx4CTq+rH3W6+BpwD\nfDzJbkn2BV4PHFlVV3Zt3g/cJslbktw+yfNo57i9Y9MdrSRJkiRt2JwHNeA5wDbAN2mThQxuj+/W\nX0G7vtoJtN6xtwL/D3jEYAdVtQbYD7gaOB04GvgocMhQm/OBh3X7+gFtWv5nVNXoTJCSJEmSNKfm\n/By1qpo0LFbVb4D7T2E/v6aFtcnanEy7PpskSZIk9VYfetQkSZIkSUMMapIkSZLUMwY1SZIkSeoZ\ng5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJMkSZKknjGoSZIkSVLPGNQkSZIkqWcMapIkSZLUMwY1\nSZIkSeoZg5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJMkSZKknjGoSZIkSVLPGNQkSZIkqWcMapIk\nSZLUMwY1SZIkSeoZg5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJMkSZKknjGoSZIkSVLPGNQkSZIk\nqWcMapIkSZLUMwY1SZIkSeoZg5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJMkSZKknjGoSZIkSVLP\nGNQkSZIkqWcMapIkSZLUMwY1SZIkSeoZg5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJMkSZKknjGo\nSZIkSVLPGNQkSZIkqWcMapIkSZLUMwY1SZIkSeoZg5okSZIk9YxBTZIkSZJ6xqAmSZIkST1jUJMk\nSZKknjGoSZIkSVLPGNQkSZIkqWcMapIkSZLUMwY1SZIkSeoZg5okSZIk9YxBTZIkSZJ6Zs6DWpJX\nJjkjycVJLkzy2SS7jLTZMsl7k6xK8tckxybZfqTNzZN8OcklSVYmOTzJZiNt7p/kzCSXJflpkqds\nimOUJEmSpOmY86AG7AW8B7gn8CBgC+BrSW4w1OadwMOAxwD3BW4GfGawsgtkxwPXA/YAngI8FThs\nqM0tgS8BJwF3Ad4FfCjJg2flqCRJkiRphq431wVU1UOH7yd5KvAHYDFwWpJtgKcDT6yqk7s2TwOW\nJ7lHVZ0B7AvcAXhAVa0Czk7yGuDNSQ6tqquA5wK/rKqXdQ+1IsmewMHA12f9QCVJkiRpivrQozbq\nxkABf+ruL6YFypMGDapqBXABcK9u0R7A2V1IGzgB2Ba401CbE0ce64ShfUiSJElSL/QqqCUJbZjj\naVV1Trd4R+CKqrp4pPmF3bpBmwsnWM8U2myTZMuNrV2SJEmSxmXOhz6OeB9wR2DPKbQNredtQyZr\nkym0AY4Dzh1ZtqS7SZIkSbouWbp0KUuXLl1r2erVqzdpDb0JakmOBB4K7FVVvxtatRK4fpJtRnrV\ntufaHrKVwN1HdrnD0LrBvzuMtNkeuLiqrpi8uv2Bw6dwFJIkSZLmuyVLlrBkydqdMsuWLWPx4sWb\nrIZeDH3sQtojaZOBXDCy+kzgKmDvofa7ADsDp3eLvg3cOcmioe32AVYDy4fa7M3a9umWS5IkSVJv\nzHmPWpL30cYQPgK4JMmg12t1VV1WVRcnOQp4R5I/A38F3g18q6q+17X9GnAO8PEkLwd2Al4PHFlV\nV3Zt3g+8IMlbgA/TQttjab14kiRJktQbfehRew6wDfBN4HdDt8cPtTmYdg20Y4faPWawsqrWAPsB\nV9N62Y4GPgocMtTmfNq12B4E/KDb5zOqanQmSEmSJEmaU3Peo1ZVGwyLVXU5cFB3W1+bX9PC2mT7\nOZk23b8kSZIk9VYfetQkSZIkSUMMapIkSZLUM9MOaklukGTrofu3SPKiJPuMtzRJkiRJWphm0qP2\neeDJAEluDHwXeAnw+STPHWNtkiRJkrQgzSSo7Q6c2v33Y2kXnb4FLby9cEx1SZIkSdKCNZOgtjXt\nWmbQLhh9XDc9/ndogU2SJEmStBFmEtR+Djwqyc2BfWkXmwbYHrh4XIVJkiRJ0kI1k6B2GPA24Hzg\nu1X17W75PsBZY6pLkiRJkhasaV/wuqqOTXIasBPww6FVJwGfHVdhkiRJkrRQTTuoAVTVSmDlyLIz\nxlKRJEmSJC1wUw5qSXYCXlBVr+run0abWGTgKuDRVfXb8ZYoSZIkSQvLdM5Rex6w3dD9u9Cm6f98\nd1sDHDxQ0Ll9AAAgAElEQVS+0iRJkiRpYZrO0Mf9WPc6ae+qql8CJPkO8A7gpWOqTZIkSZIWpOn0\nqN0SOG/o/teBS4burwBuNYaaJEmSJGlBm06P2hbAPwK/Aaiq/UfWb0cb/qhZ9ve//51ly5bNdRnX\nWLRoETvvvPNclyFJkiRdZ0wnqK0A7s36r5W2F/DTja5IG3ApJ554Kl/72uK5LuQaW221NStWLDes\nSZIkSWMynaD2KeCwJKdW1Y+GVyS5C3AI8OZxFqeJXMGaNVcCxwC7znUxwHIuu+xAVq1aZVCTJEmS\nxmQ6Qe2dtAlFzkzydVoPWwF3AB4MfLtro01iV2D3uS5CkiRJ0iyYclCrqiuTPBh4MfBE4P7dqp8B\nrwGOqKorx16hJEmSJC0w0+lRo6quoA1vdIijJEmSJM2SaQU1gCQ3oA113IU29PGnwIlVdemYa5Mk\nSZKkBWlaQS3JI4APAYtGVq1K8oyq+uLYKpMkSZKkBWrKF7xOcm/gWOAU4D7ATbrbnsCpwLFJ7jUb\nRUqSJEnSQjKdHrVXAx+pqmePLD8dOD3JB2iTijx0XMVJkiRJ0kI05R41YA/gyEnWvxewR02SJEmS\nNtJ0gtoNgIsnWb8a2GrjypEkSZIkTSeo/Qx44CTr9+7aSJIkSZI2wnSC2keAtyVZ5xy0JA8DDgc+\nOqa6JEmSJGnBms5kIu8C7g18KckKYHm3/I7A7YDPAe8cb3mSJEmStPBMuUetqtZU1eOAJcAK4A7d\n7VzggKp6TFWtmZ0yJUmSJGnhmNYFrwGq6n+A/5mFWiRJkiRJTCOoJdlmKu2qarKZISVJkiRJGzCd\nHrW/ADXJ+nTrN9+oiiRJkiRpgZtOUHvA0H8HOB54JvDbsVYkSZIkSQvclINaVZ08fD/J1cB3quqX\nY69KkiRJkhaw6VxHTZIkSZK0CRjUJEmSJKlnNjaoTTa5iCRJkiRpBqYzPf9xI4u2At6f5JLhhVW1\n/zgKkyRJkqSFajqzPq4euX/MOAuRJEmSJDXTmfXxabNZiCRJkiSpcTIRSZIkSeoZg5okSZIk9YxB\nTZIkSZJ6xqAmSZIkST0zpaCWZFmS7br/fm2SrWe3LEmSJElauKbao7YrcMPuvw8B/mF2ypEkSZIk\nTXV6/h8AH0lyGhDgpUn+NlHDqjpsXMVJkiRJ0kI01aD2VOB1wH5AAQ8BrpqgXQEGNUmSJEnaCFMK\nalW1AngiQJI1wN5V9YfZLEySJEmSFqqp9qhdo6qcKVKSJEmSZtG0gxpAktsAL6JNMlLAcuBdVfWL\nMdYmSZIkSQvStHvHkuwLnAPcA/gR8GPgnsBPkjx4vOVJkiRJ0sIzkx61NwNHVNUrhhcmeTPwFuDr\n4yhMkiRJkhaqmZxvtitw1ATLPwzccSZFJNkryReS/DbJmiSPGFn/kW758O34kTbbJflEktVJ/pzk\nQ0luONJmtySnJLk0ya+S/MdM6pUkSZKk2TSToHYR8K8TLP9XYKYzQd6Qdq2259POeZvIV4AdgB27\n25KR9Z+khci9gYcB9wU+MFiZ5EbACcB5wO7AfwCHJnnmDGuWJEmSpFkxk6GPHwT+O8mtgdNpwWpP\n4OXA22dSRFV9FfgqQJKsp9nlVXXRRCuS3AHYF1hcVWd1yw4CvpzkpVW1EjgQ2AJ4RlVdBSxPclfg\nxcCHZlK3JEmSJM2GmfSovZ52UeuDgJOBU4AXAIcCbxhbZeu6f5ILk5yb5H1JbjK07l7AnwchrXMi\nLUTes7u/B3BKF9IGTgBun2TbWaxbkiRJkqZlJtdRK+AI4IhuOCFV9ddxFzbiK8BnaMMWbwO8CTg+\nyb26enZkZNhlVV2d5E/dOrp/fzmy3wuH1q2epdolSZIkaVpmdB21gU0Q0AaP8+mhuz9JcjbwC+D+\nwDcm2TSs/5y3wXo20AY4Djh3ZNkS1j1NTpIkSdJ8t3TpUpYuXbrWstWrN22/zkYFtblSVeclWQXc\nlhbUVgLbD7dJsjmwXbeO7t8dRnY12OZCJrU/cPhG1SxJkiRpfliyZAlLlqzdKbNs2TIWL168yWqY\nyTlqcy7JPwM3BX7fLfo2cONucpCBvWk9ZmcMtblvF+AG9gFWVJXDHiVJkiT1Ri+CWpIbJrlLksG0\n/7fu7t+8W3d4knsmuUWSvYHPAT+lTQZCVZ3b/fcHk9w9yX2A9wBLuxkfoU3ffwXw4SR3TPIE4IXM\ncKZKSZIkSZot0wpqSbZIclKS2425jrsBZwFn0s4XezuwDHgdcDWwG/B5YAXt8gDfA+5bVVcO7eNJ\ntBPJTgS+RJuN8tmDlVV1MW0K/1sC3wfeChxaVRNdvFuSJEmS5sy0zlGrqiuT7DbuIqrqZCYPjf82\nhX38hXattMnanA3cb3rVSZIkSdKmNZOhj8cAzxh3IZIkSZKkZiazPl4PeHqSB9OGEF4yvLKqXjyO\nwiRJkiRpoZpJUPsX2vljALuMrNvA9cgkSZIkSRsy7aBWVQ+YjUIkSZIkSc2Mp+dPctsk+ya5QXc/\n4ytLkiRJkhauaQe1JDdNchLtOmbHAzt1q45K4jXJJEmSJGkjzaRH7QjgSmBn4O9Dy/+HKUyjL0mS\nJEma3EwmE9kH2LeqfjMy2vFnwC3GUpUkSZIkLWAz6VG7IWv3pA3cBLh848qRJEmSJM0kqJ0KPHno\nfiXZDHgZ8I2xVCVJkiRJC9hMhj6+DDgpyd2A6wOHA3ei9ajdZ4y1SZIkSdKCNO0etar6Me1C16cB\nn6cNhTwOuGtV/WK85UmSJEnSwjOTHjWqajXwhjHXIkmSJElihkEtyXbAM4BdgQKWAx+pqj+NsTZJ\nkiRJWpBmcsHr+wLnAy8EtqOdm/ZC4LxunSRJkiRpI8ykR+29tItbP7eqrgZIsjnwvm7dncdXniRJ\nkiQtPDOZnv+2wNsHIQ2g++93dOskSZIkSRthJkFtGe3ctFG7Aj/cuHIkSZIkSVMa+phkt6G77wbe\nleS2wHe6ZXsAzwdeMd7yJEmSJGnhmeo5aj+gze6YoWWHT9Duk7Tz1yRJkiRJMzTVoHarWa1CkiRJ\nknSNKQW1qvrVbBciSZIkSWpmesHrmwF7AtszMiFJVb17DHVJkiRJ0oI17aCW5KnAB4ArgD/Szl0b\nKNpkI5IkSZKkGZpJj9rrgcOAN1XVmjHXI0mSJEkL3kyuo7Y18ClDmiRJkiTNjpkEtaOAx427EEmS\nJElSM5Ohj68EvpTk34CzgSuHV1bVi8dRmCRJkiQtVDMNavsCK7r7o5OJSJIkSZI2wkyC2kuAp1fV\nR8dciyRJkiSJmZ2jdjnwrXEXIkmSJElqZhLU3gUclCTjLkaSJEmSNLOhj/cAHgjsl+QnrDuZyP7j\nKEySJEmSFqqZBLW/AMeNuxBJkiRJUjPtoFZVT5uNQiRJkiRJzUzOUZMkSZIkzaJp96glOY9JrpdW\nVbfeqIokSZIkaYGbyTlq7xy5vwVwV+DfgLdudEWSJEmStMDN5By1d020PMnzgbttdEWSJEmStMCN\n8xy1rwCPGeP+JEmSJGlBGmdQeyzwpzHuT5IkSZIWpJlMJnIWa08mEmBH4B+B542pLkmSJElasGYy\nmcjnRu6vAS4CvllV5258SZIkSZK0sM1kMpHXzUYhkiRJkqTGC15LkiRJUs9MuUctyRomudB1p6pq\nJsMpJUmSJEmd6YSqR0+y7t7AQbSJRSRJkiRJG2HKQa2qPj+6LMkdgDcBDwc+AbxmfKVJkiRJ0sI0\no3PUktwsyQeBH9HC3l2r6ilVdcFYq5MkSZKkBWhaQS3JtkneAvwcuBOwd1U9vKrOnpXqJEmSJGkB\nms5kIi8DXg6sBJZMNBRSkiRJkrTxpjOZyJuBS2m9aU9J8pSJGlXV/uMoTJIkSZIWqukEtaPZ8PT8\nkiRJkqSNNJ1ZH586W0Uk2Qv4D2AxsBPwqKr6wkibw4BnAjcGvgU8t6p+PrR+O+BIYD9gDfAZ4N+r\n6pKhNrt1be4O/AE4sqreOlvHJUmSJEkzMaNZH2fBDYEfAM9ngl67JC8HXgA8G7gHcAlwQpLrDzX7\nJLArsDfwMOC+wAeG9nEj4ATgPGB3WjA8NMkzZ+F4JEmSJGnGpjP0cdZU1VeBrwIkmeii2f8OvL6q\nvti1eTJwIfAo4NNJdgX2BRZX1Vldm4OALyd5aVWtBA4EtgCeUVVXAcuT3BV4MfChWT1ASZIkSZqG\nvvSorVeSWwE7AicNllXVxcB3gXt1i/YA/jwIaZ0Tab1z9xxqc0oX0gZOAG6fZNtZKl+SJEmSpq33\nQY0W0orWgzbswm7doM0fhldW1dXAn0baTLQPhtpIkiRJ0pzrxdDHGQobnoVyQ20Gwyw3sJ/jgHNH\nli3pbpIkSZKuS5YuXcrSpUvXWrZ69epNWsN8CGoraYFqB9buEdseOGuozfbDGyXZHNiuWzdos8PI\nvgfbjPa0jdgfOHx6VUuSJEmal5YsWcKSJWt3yixbtozFixdvshp6P/Sxqs6jhay9B8uSbEM79+z0\nbtG3gRt3k4MM7E0LeGcMtblvF+AG9gFWVNWmjceSJEmSNIleBLUkN0xylyT/2i26dXf/5t39dwKv\nTvLwJHemXXz7N8DnAarqXNrEIB9Mcvck9wHeAyztZnyENn3/FcCHk9wxyROAFwJv3yQHKUmSJElT\n1Jehj3cDvkE7V6y4Njx9DHh6VR2eZGvaddFuDJwKPKSqrhjax5NoF7M+kXbB62Np0/oDbabIJPt2\nbb4PrAIOraqjZvPAJEmSJGm6ehHUqupkNtC7V1WHAodOsv4vtGulTbaPs4H7Tb9CSZIkSdp0ejH0\nUZIkSZJ0LYOaJEmSJPWMQU2SJEmSesagJkmSJEk9Y1CTJEmSpJ4xqEmSJElSzxjUJEmSJKlnDGqS\nJEmS1DMGNUmSJEnqGYOaJEmSJPWMQU2SJEmSesagJkmSJEk9Y1CTJEmSpJ4xqEmSJElSzxjUJEmS\nJKlnDGqSJEmS1DMGNUmSJEnqGYOaJEmSJPWMQU2SJEmSesagJkmSJEk9Y1CTJEmSpJ4xqEmSJElS\nzxjUJEmSJKlnDGqSJEmS1DMGNUmSJEnqGYOaJEmSJPWMQU2SJEmSesagJkmSJEk9Y1CTJEmSpJ4x\nqEmSJElSzxjUJEmSJKlnDGqSJEmS1DMGNUmSJEnqGYOaJEmSJPWMQU2SJEmSesagJkmSJEk9Y1CT\nJEmSpJ4xqEmSJElSzxjUJEmSJKlnDGqSJEmS1DMGNUmSJEnqGYOaJEmSJPWMQU2SJEmSesagJkmS\nJEk9Y1CTJEmSpJ4xqEmSJElSzxjUJEmSJKlnDGqSJEmS1DPXm+sCdN2wfPnyuS5hLYsWLWLnnXee\n6zIkSZKkGTGoaSP9HtiMAw88cK4LWctWW23NihXLDWuSJEmalwxq2kh/AdYAxwC7znEtA8u57LID\nWbVqlUFNkiRJ85JBTWOyK7D7XBchSZIkXSc4mYgkSZIk9YxBTZIkSZJ6Zl4EtSSHJFkzcjtnaP2W\nSd6bZFWSvyY5Nsn2I/u4eZIvJ7kkycokhyeZF8cvSZIkaWGZT+eo/RjYG0h3/6qhde8EHgI8BrgY\neC/wGWAvgC6QHQ/8DtgDuBnwceAK4NWboHZJkiRJmrL5FNSuqqqLRhcm2QZ4OvDEqjq5W/Y0YHmS\ne1TVGcC+wB2AB1TVKuDsJK8B3pzk0Kq6anS/kiRJkjRX5tPQv9sl+W2SXyQ5JsnNu+WLaYHzpEHD\nqloBXADcq1u0B3B2F9IGTgC2Be40+6VLkiRJ0tTNl6D2HeCptJ6x5wC3Ak5JckNgR+CKqrp4ZJsL\nu3V0/144wXqG2kiSJElSL8yLoY9VdcLQ3R8nOQP4FfB44LL1bBagprL7DTc5Djh3ZNmS7iZJkiTp\numTp0qUsXbp0rWWrV6/epDXMi6A2qqpWJ/kpcFvgROD6SbYZ6VXbnmt7zVYCdx/ZzQ7dv6M9bRPY\nHzh8Y0qWJEmSNE8sWbKEJUvW7pRZtmwZixcv3mQ1zJehj2tJ8g/AbWizOJ5JmwFy76H1uwA7A6d3\ni74N3DnJoqHd7AOsBs5BkiRJknpkXvSoJXkr8EXacMd/Al5HC2efqqqLkxwFvCPJn4G/Au8GvlVV\n3+t28TVaIPt4kpcDOwGvB46sqis37dFIkiRJ0uTmRVAD/hn4JHBT4CLgNGCPqvpjt/5g4GrgWGBL\n4KvA8wcbV9WaJPsB/0XrZbsE+ChwyCaqX5IkSZKmbF4EtaqadNaOqrocOKi7ra/Nr4H9xlyaJEmS\nJI3dvDxHTZIkSZKuywxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrGoCZJkiRJPWNQkyRJ\nkqSeMahJkiRJUs8Y1CRJkiSpZwxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrGoCZJkiRJ\nPWNQkyRJkqSeMahJkiRJUs8Y1CRJkiSpZwxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFNkiRJknrG\noCZJkiRJPWNQkyRJkqSeMahJkiRJUs8Y1CRJkiSpZwxqkiRJktQzBjVJkiRJ6hmDmiRJkiT1jEFN\nkiRJknrGoCZJkiRJPWNQkyRJkqSeMahJkiRJUs8Y1CRJkiSpZwxqkiRJktQz15vrAqTZsnz58rku\n4RqLFi1i5513nusyJEmSNE8Y1HQd9HtgMw488MC5LuQaW221NStWLDesSZIkaUoMaroO+guwBjgG\n2HWOawFYzmWXHciqVasMapIkSZoSg5quw3YFdp/rIiRJkqRpczIRSZIkSeoZg5okSZIk9YxBTZIk\nSZJ6xqAmSZIkST1jUJMkSZKknjGoSZIkSVLPGNQkSZIkqWcMapIkSZLUMwY1SZIkSeoZg5okSZIk\n9YxBTZIkSZJ65npzXYC0UCxfvnyuS7jGokWL2Hnnnee6DEmSJK2HQU2adb8HNuPAAw+c60KusdVW\nW7NixXLDmiRJUk8Z1KRZ9xdgDXAMsOsc1wKwnMsuO5BVq1YZ1CRJknrKoCZtMrsCu891EZIkSZoH\nFtxkIkmen+S8JJcm+U6Su891TZKutXTp0rkuQTPkazc/+brNX75285evnaZiQfWoJXkC8HbgWcAZ\nwMHACUl2qapVc1qctIn1dXKTpUuXsmTJkjmuSDPhazc/+brNX75285evnaZiQQU1WjD7QFUdDZDk\nOcDDgKcDh89lYdKm07/JTbbccis+85lj2WmnnVi9ejXLli2b65KcGVOSJM2pBRPUkmwBLAbeOFhW\nVZXkROBec1aYtMn1bXKTU7n88hez3377XbNk8eLFc1hPMxwe+8DgKEnSwrJgghqwCNgcuHBk+YXA\n7Sff9EJg7n/hb/421wXoOqMvk5ssZ+3geDBwxJxWNFF4nGt9C46XX345W2655VrL5rI3dKJ65lrf\najLsT+6CCy5g1ap+nQWxoffQpv7M+R6af/r2vvY9ND0LKaitT4Baz7qt2j9Hd7c+OZ72BXeufav7\nty/1QP9qsp7JDeo5r/v3r8x9XSto4fEZQB+C0c+4/PJP9yo4trmo1qyzdO56QyeuZ271q6YtttiS\nt771LSxatGit5b/5zW/4xCc+scnr2WyzzVizph/Pz6pVq/iP/3gFV1552VyXMmLD76FN+Zlb33to\nrvTpPQTTq2dTfO76+L6+/vW34rjj+vOj43QNnd+/1aZ4vFStL6Nct3RDH/8OPKaqvjC0/KPAtlX1\n6Am2eRKw6f/vJUmSJKmvDqiqT872gyyYHrWqujLJmcDewBcAkqS7/+71bHYCcABwPtCfnyMkSZIk\nbWpbAbekZYRZt2B61ACSPB74GPBsrp2e/7HAHarqormsTZIkSZIGFkyPGkBVfTrJIuAwYAfgB8C+\nhjRJkiRJfbKgetQkSZIkaT7YbK4LkCRJkiStzaAmSZIkST1jUFuPJM9Pcl6SS5N8J8nd57qmhSzJ\nK5OckeTiJBcm+WySXUbabJnkvUlWJflrkmOTbD/S5uZJvpzkkiQrkxyexM/BJtS9lmuSvGNoma9d\nTyW5WZKPd6/N35P8MMnuI20OS/K7bv3Xk9x2ZP12ST6RZHWSPyf5UJIbbtojWTiSbPb/t3fnUVaU\nZx7Hvz9FjMaFJC7ESIy4gBpFxAWNSyIRRgwajxM0cxx0ohPHJW6ZIHo0xphRNAY1ahJPjAtGY2I4\ng+IYF9SJBlGPQNQRRRRcEdyQgCvCM3+879Wi+nb3Rbr73u7+fc6p011V731refrerqfet94r6VxJ\nc3JMnpV0ZpVyjludSdpL0q2SXsmfiwdWKbPKcZK0g6T78zXNC5J+1N7H1tW1FDtJPSRdIOlxSUty\nmeskfbFUh2NXB7W87wplr8xlTiwt75DY+SKnCkmHAr8AzgYGAo8BdyoNRGL1sRdwGbAb8E1gDeAu\nSWsVylwCHAAcAuwNbAJMqKzMF/W3kwbRGQwcARxJGlzGOoDSDY9/J72nihy7BiSpF+kbyT8AhgHb\nAD8EFhbKnAacQBpNd1fgHdLnZc9CVTfm1w4hxXlv4MoOOITuagwpHscB/YHRwGhJJ1QKOG4N47Ok\ngc2OB5oMGtAWcZK0Lmko8bnATsCPgJ9IOrodjqc7aSl2awM7AueQriMPBvoBt5TKOXb10eL7rkLS\nt0nvu1eqrO6Y2EWEp9IEPARcWpgX8DIwut775unjmGwALAf2zPPrkS4mDy6U6ZfL7Jrn9weWAhsU\nyhxDuujsUe9j6uoTsA4wC9gXuA8Y59g19gSMBf7aSpl5wCmF+fWA94CReX6bHMuBhTLDgI+A3vU+\nxq44AZOA35aW/RkY77g17pTP94GlZascJ+BY4I3iZyVwPjCz3sfcVaZqsatSZmdgGbCpY9c4U3Ox\nA74EvJjjNBc4sbCuf0fFzi1qJZLWAAYB91SWRTq7k4Hd67Vf1kQv0l2Qt/L8IFJrSzFus0hvskrc\nBgNPRMQbhXruBNYHtmvvHTauACZFxL2l5Tvj2DWqEcCjkv6k1OV4evFuoKTNgd6sGLt/AA+zYuwW\nRsSMQr2TSe/f3dr7ALqpB4EhkrYCkDQA+BqpVdpx6yTaME6Dgfsj4qNCmTuBfpLWb6fdt6Yq1y1v\n53nHrkFJEjAeuDAinqpSZHc6KHZO1JraAFgdWFBavoD0gWl1lt9AlwB/i4iZeXFv4MP8T6yoGLfe\nVI8rOLbtStJhpO4fp1dZvTGOXaPqS7orOAsYCvwG+KWkw/P63qR/TC19XvYGXiuujIhlpJssjl37\nGAv8EXha0ofANOCSiLgpr3fcOoe2ipM/P+tM0pqk9+WNEbEkL3bsGtcY0nXJ5c2s77DYdasvvF5F\nooV+rNahfgVsC+xZQ9la4+bYthNJm5IS6/0iYunKvBTHrt5WAx6JiLPy/GOStiMlb79v4XW1xM6f\nqe3nUOBfgMOAmaRnZS6VNC8irm/hdY5b59AWcVL+6Vi2M0k9gJtJ5/q4Wl6CY1c3kgYBJ5JuLq/0\ny2nj2LlFrak3SH2INy4t34immbF1MEmXA8OBr0fEvMKq+UBPSeuVXlKM23yaxrUy79i2n0HAhsA0\nSUslLQX2AU7Kd/sXAGs6dg3pVaDc7eMp4Mv59/mkfzwtfV7Oz/Mfk7Q68Dkcu/ZyIXB+RNwcEU9G\nxA3AxXzSou24dQ6rGqf5hTLV6gDHsl0VkrQ+wNBCaxo4do1qT9I1y0uFa5bNgHGS5uQyHRY7J2ol\n+Y7/NNIoLsDHXe2GkPr9W53kJO0g4BsR8WJp9TTSQ5zFuG1NuqCsxG0qsH1p9M6hwCLSXWdrH5OB\n7Ul39Qfk6VFSi0zl96U4do1oCmlgl6J+wAsAETGX9M+oGLv1SH30i7HrJal4d3II6QL04fbZ7W5v\nbZresV1O/p/vuHUObRCnRwpl9s4XkhVDgVkRsaiddr/bKyRpfYEhEbGwVMSxa0zjgR345HplAGlQ\nnwtJA4ZAR8au3qOtNOIEjCSNqjSKNLLLlcCbwIb13rfuOpG6Oy4kDdO/cWH6TKnMXODrpFacKcAD\nhfWrkYaF/0t+Ew4j3dU4t97H190mCqM+OnaNO5EGevmA1BKzBak73WLgsEKZ0fnzcQQpIZ8IzAZ6\nFsrcTkrIdyENajELuL7ex9dVJ+Aa0mA8w0l3gg8mPU9xnuPWWBNpmPABpBtZy4GT83yftooTaaTI\necB1pMcGDgWWAEfV+/g789RS7EhjHdxCuqm1fem6ZQ3HrnFj10z5FUZ97MjY1f1kNepE6kf8PClh\nmwrsXO996s5TfiMtqzKNKpRZk/Rda2+QLiZvBjYq1dMHuC2/WRYAFwCr1fv4utsE3MuKiZpj16AT\n6WL/ceBd4Enge1XK/CT/Q3qXNKrVlqX1vUgtqItIN1x+C6xd72PrqlO+CBmXLy7eIV3Yn0Ppqywc\nt/pPpG7g1f6/Xd2WcSIlC3/NdbwI/Ge9j72zTy3FjnSDpLyuMr+3Y9e4sWum/ByaJmodEjvliszM\nzMzMzKxB+Bk1MzMzMzOzBuNEzczMzMzMrME4UTMzMzMzM2swTtTMzMzMzMwajBM1MzMzMzOzBuNE\nzczMzMzMrME4UTMzMzMzM2swTtTMzMzMzMwajBM1MzMzMzOzBuNEzczMVpqkzSQtl7RDvfelQlI/\nSVMlvSdper33x8zMbFU4UTMz64QkXZsTpdGl5QdJWt5BuxEdtJ1anQMsAbYChjRXSNLGki6VNDsn\nda9Kul/SMZLW6rC9bXAtJeOS7pM0rrRsW0l/lPSapPclzZJ0TrVzKmmgpJskzctl50q6VdK32vOY\nzMw6EydqZmadUwDvAadJWr/Kuo6gNq9QWmMVXr4F8LeIeDkiFjZT/+bA34FvAmOAHYHdgZ8D36KF\nBK+rkrSapOZiWdPfkqTBwMNAD2B/UrJ8BnAkcLekHoWyBwFTgbWBUUB/4J+AicC5ktb7dEdiZta1\nOFEzM+u8JgPzSRfEVUk6W9KM0rKTJM0tzF8j6b8lnS5pvqSFks6UtLqkCyW9KeklSUdW2cQ2kqbk\nli6Zjc4AAAdYSURBVKknJO1d2tZXJd0uaXGue7ykLxTW3yfpMkkXS3oduKOZ45CkH+f9eF/SDEnD\nCuuXAzsBZ0taJunHzZySXwMfAoMiYkJEzIqI5yNiUkSMiIjbCnWuL+mq3EK0SNLkYutS5dxKOjy3\nCL0t6Q+SPlva79MlzZH0bi5/SGF9L0k35G28m1uhjmhm34vn67K8vdcl/bRUpqekiyS9LGlJ7g66\nT2H9ETnGIyQ9CbwP9Gluk83tS8nvgCcj4pCImBYRL0XEBGAEKRE+JW97beAqYFJEHBgRk/P5nxUR\nV0fEwIj4R43bNDPr0pyomZl1XstISdoPJG3SQrlqrSLlZfsCXwT2Il1U/xS4DXgL2BX4DXBlle1c\nSGqN2pHUSjJJ0ucgJTrAPcA0UhI1DNgI+FOpjlHAB8AewH80cwwn5/06FdgeuBO4VdIWeX1vYCZw\nUT6Oi8oVSPo8sB9weUS838x2iv4MfCHv907AdGCypF6FMlsABwHDgQOAfUgtdRVnAIcD3we2BS4G\nrpe0V17/M1KL0rD881jgjVb2axSwFNgFOBE4VdJRhfVXALsBI0nn6mbgL4VzBak1azRwFLAd8For\n22yWpIHANsC48rqIeJx0Q+G7edEw4POkvxszM2uBEzUzs04sIm4hdeU7ZxWrehM4KSJmR8S1wCxg\nrYgYGxHPAeeTWqL2LL3usoiYGBGzSEnGItLFP8AJwPSIOCvX+xhwNPANSVsW6ng2IsbkMrOb2b8f\nAmMj4uZcbkw+7pMBIuI14CNgSUS8FhHvVqljS1IL0TPFhblVanGezs/L9gR2BkZGxIyIeC4iRufj\n++fiy4EjIuKpiJgCXE/uPimpJ3A68L1Cy9F44AbgmPz6PsCMvI0XI+LeiPifZs5BxUsRcWo+D38A\nLuOTFqsvk7obficiHoyIuRExDpgC/Fuhjh7AsRHxUK6npcT1wcL5WSxpMSv+HWxFSvyfbub1TwFb\nF8pCIQaSdi7VP7yV4zcz6xZ6tF7EzMwa3GnAPZJ+sQp1PBkRxVa2BcATlZmIWC7pTVKLWNFDhTLL\nJD1Kal0BGADsmy/si4LUEvVsnn+0pR2TtC6wCfBgadUU4NOMOlluTdyFdOPyRmDNvGwHYF3grdLj\nW58h7XvF86Wk8FU+OUdbklqu7i49A7YGqXUOUlfMCZIGAXcBEyNiaiv7/1BpfiqpVU3AV4HVgWdK\n2+zJii11H0bE/7WynYqRNE3CbqzxtZCS2ZaedXuM9LcC6W/C1yZmZvjD0Mys04uIByTdSWr1ura0\nejlNnzOqNmDH0nK1zSyrpSdG5aJ8HeBWUhe78j68Wvj9nRrqLNZb0VoCUPZsLt8/71eqNOJ5AEnv\nFcquA8wjdWUs7/vbhd9bOkfr5J/Dc11FH+Rt35FbwQ4gDXByj6TLc+vdp7EOqWVxJ1Lsi5YUfn+P\n2r0cEXOKC0rn6hnSOdoWeLzK6/sDlZbSys9+wCMAEbEUmJPrXYndMjPr2tz10cysaziDNHDDHqXl\nr5Oe3yoa2IbbHVz5RdLqwCBSVzdIrUbbAS9ExJzSVHOiEBGLSYlOudvlHoVt1VLPW8DdwAlqfRj+\n6aTztqzKvr9V4yZnkhKyzarU8Uphv96MiPERMYrUlfP7rdQ7uDS/OzA7t4jOILWobVxlm5/mObRW\nE+GI+Dupxe2U8jpJA0gJaKUF7i5gIakV2MzMWuBEzcysC4iIJ0jPPv2gtOp/gQ0ljZbUV9LxpKHQ\n28rxkr4tqR/wK6AXcE1edwVp4Iib8nNIfSUNk3S1Vr7p5OekryIYKWlrSWNJ3eUuXcl6jiP1Jnk0\n19U/13c4qeVnGUBETCZ1KZwoaT+l7xTbQ9LPJO1Uy4YiYglpUJOLJY3Kxz9Q0gmS/hVA6XvGDpS0\nhaTtSF8RMLOVqvvkUR23lvRd0rOAl+RtziYlReMlHSzpK5J2lTRG0v4rea6g9lEfjwa2lTRB0i6S\n+kj6Dqnlcgo5ThHxTi57gKTbJA2VtLmk7ZW+EzDIMTAz6+7c9dHMrOs4i/Q80cetIBHxtKTjSC1u\nZwITSElPa602tYwUGaQRDseQkqZngRGVFqeIeFXS14ALSKM0rgm8ANxReB6u1q6LvyQ9M3YR6Rmw\nmXlbz7WyzyvucMScPErhGcB5wKakVq+ZpJEIf10oPhz4L+BqYEPSVyHcT3p+ryYRcZakBaRz1JfU\nbXJ63jakAVrOA75C6o74AJ+MkNic8cBapK6DHwEXR8RVhfVHkmJ9EfAl0kAxU4FJte538RBqWR4R\nDyp9l9rZwO2kWL1IStrH5u6NlbITJe1BalW7jpTMLyI9q3hoDYOpmJl1C1rx2XEzMzNrVJLuI40S\neWq998XMzNqXuz6amZmZmZk1GCdqZmZmnYe7wZiZdRPu+mhmZmZmZtZg3KJmZmZmZmbWYJyomZmZ\nmZmZNRgnamZmZmZmZg3GiZqZmZmZmVmDcaJmZmZmZmbWYJyomZmZmZmZNRgnamZmZmZmZg3GiZqZ\nmZmZmVmD+X8GQLcK5JwDjQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x4d2dbe0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%matplotlib inline\n",
    "import pandas\n",
    "lengths = [sum(len(OGs[label]) for label in group) for group in smart_modified_HOGs]\n",
    "df = pandas.Series(lengths)#, columns=['Family ID', 'Copies'])\n",
    "bins = 20\n",
    "f = df.hist(bins=bins, figsize=(10,5), )#dpi=300)\n",
    "f.grid(False)\n",
    "f.set_title(\"Did we damp down on extremely large HOGs?\")\n",
    "f.set_xlabel(\"Number of Genes per HOG\")\n",
    "f.set_ylabel(\"Number of HOGs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Conclusion: Just filter out HOGs with > 500 genes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-15'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_smart_modified_HOGs = [group for group in smart_modified_HOGs if sum(len(OGs[label]) for label in group) < 500]\n",
    "pp(len(filtered_smart_modified_HOGs) - len(smart_modified_HOGs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import pandas\n",
    "lengths = [sum(len(HOGs[label]) for label in group) for group in modified_HOGs_only]\n",
    "df = pandas.Series(lengths)#, columns=['Family ID', 'Copies'])\n",
    "f = df.hist(bins=100, figsize=(10,5), )#dpi=300)\n",
    "f.grid(False)\n",
    "f.set_title(\"Did we damp down on extremely large HOGs?\")\n",
    "f.set_xlabel(\"Number of Genes per HOG\")\n",
    "f.set_ylabel(\"Number of HOGs\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Concat Super HOG fasta into new directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from os.path import join, basename, splitext\n",
    "import os\n",
    "family_out_dir = r\"D:\\Genomes\\Ash_Gene_Families\\capped\"\n",
    "HOGfasta = r\"D:\\josiah\\Documents\\Research\\Thesis - Genome Symmetry\\DNA_Duplications\\Ash_Proteome\\Results_Jun25\\Orthologues_Jul06\\Sequences\"\n",
    "os.makedirs(family_out_dir, exist_ok=True)\n",
    "for family in filtered_smart_modified_HOGs:\n",
    "    family_name = ('-'.join(family))[:200] + '.fa'\n",
    "    with open(join(family_out_dir, family_name), 'wb') as concat:  # output file concatenation\n",
    "        for hog in family:\n",
    "            with open(join(HOGfasta, hog + '.fa'), 'rb') as hog_file:  # smaller input file\n",
    "                concat.write(hog_file.read())  # dump the whole file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pp(len(filtered_smart_modified_HOGs))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.4.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
